{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TODO\n",
    "* `gb_accession` and `gisaid_accession` are not found for new sequences, how do we concat to `metadata.csv` without them?\n",
    "* metadata format for NCBI\n",
    "* support tools for manual sanity checks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "from bjorn import *\n",
    "from bjorn_support import *\n",
    "import gffutils"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Retrospect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "# * difference in sample 4459 at position 14118\n",
    "# * difference in sample 4453 at position 9030\n",
    "# * difference in sample 4404 at positions 9016, 9019, 9020\n",
    "# * difference in sample 4116 at position 16708\n",
    "# * difference in sample 4387 at position 21969"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_ids = [4459, 4453, 4404, 4116, 4387]\n",
    "# new iVar samples\n",
    "new_analysis = \"/home/al/analysis/11-16.11.2020\"\n",
    "## variants\n",
    "new_variant_paths = get_filepaths(sample_ids=sample_ids, analysis_path=new_analysis, \n",
    "                                  data_fmt='tsv', data_type='variants')\n",
    "new_variant_data = get_variant_data(new_variants)\n",
    "# old iVar samples\n",
    "old_analysis = \"/home/gk/analysis/2020.11.16.hCoV19\"\n",
    "## variants\n",
    "old_variant_paths = get_filepaths(sample_ids=sample_ids, analysis_path=old_analysis, \n",
    "                                  data_fmt='tsv', data_type='variants')\n",
    "old_variant_data = get_variant_data(old_variant_paths)\n",
    "# add type col to differentiate between samples\n",
    "new_variant_data['type'] = 'new'\n",
    "old_variant_data['type'] = 'old'\n",
    "pd.concat([old_variant_data, new_variant_data]).to_csv(\"variants_of_interest.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## bams\n",
    "new_bam_paths = get_filepaths(sample_ids=sample_ids, analysis_path=new_analysis, data_fmt='bam', data_type='trimmed_bams')\n",
    "new_bam_paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "glob.glob(f\"{analysis_path}/*{s_id}*.fa*\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "387"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_cns_filepaths(sample_ids: list=[], analysis_path: str='/home/gk/analysis') -> dict:\n",
    "    \"\"\"Takes list of sample IDs and returns filepaths of their corresponding sequences\"\"\"\n",
    "    cns_paths = {}\n",
    "    if sample_ids:\n",
    "        for s_id in sample_ids:\n",
    "            f = glob.glob(f\"{analysis_path}/*{s_id}*.fa*\")\n",
    "            try:\n",
    "                cns_paths[s_id] = f[0]\n",
    "            except:\n",
    "                continue\n",
    "    else:\n",
    "        fs = glob.glob(f\"{analysis_path}/*.fa*\")\n",
    "        for f in fs:\n",
    "            try:\n",
    "                sample_id = f.split('/')[-1].split('_')[0].split('-')[1]\n",
    "            except:\n",
    "                continue\n",
    "            cns_paths[sample_id] = f\n",
    "    return cns_paths\n",
    "analysis_pth = '/home/al/analysis/11-16.11.2020/consensus_sequences/illumina'\n",
    "new_cns_filepaths = get_cns_filepaths(analysis_path=analysis_pth)\n",
    "len(new_cns_filepaths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def get_filepaths(analysis_path: str, data_fmt: str, sample_ids: list=[], data_type: str='consensus', tech: str='illumina') -> dict:\n",
    "    \"\"\"Take list of sample IDs, the general area where your files, and their format.\n",
    "    \"\"\"\n",
    "    file_paths = {}\n",
    "    if sample_ids:\n",
    "        for s_id in sample_ids:\n",
    "            f = glob.glob(f\"{analysis}/*{data_type}*/*{tech}*/*{s_id}*.{data_fmt}\")\n",
    "            try:\n",
    "                file_paths[s_id] = f[0]\n",
    "            except:\n",
    "                continue\n",
    "    else:\n",
    "        fs = glob.glob(f\"{analysis_path}/*.fa*\")\n",
    "        for f in fs:\n",
    "            try:\n",
    "                sample_id = f.split('/')[-1].split('_')[0].split('-')[1]\n",
    "            except:\n",
    "                continue\n",
    "            file_paths[sample_id] = f\n",
    "    return file_paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "243"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get cns seqs from iVar < 1.3\n",
    "analysis_pth = '/home/al/code/HCoV-19-Genomics/consensus_sequences'\n",
    "old_cns_filepaths = get_filepaths(sample_ids=new_cns_filepaths.keys(),\n",
    "                                  analysis_path=analysis_pth,\n",
    "                                  data_fmt='fa')\n",
    "len(old_cns_filepaths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " get_filepaths(sample_ids=new_cns_filepaths.keys(),\n",
    "                analysis_path=analysis_pth,\n",
    "                                  data_fmt='fa'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'4453': '/home/gk/analysis/2020.11.16.hCoV19/variants/illumina/SEARCH-4453_L001_L002_L003_L004.tsv'}"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_variant_filepaths(['4453'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>REGION</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>REF_DP</th>\n",
       "      <th>REF_RV</th>\n",
       "      <th>REF_QUAL</th>\n",
       "      <th>ALT_DP</th>\n",
       "      <th>ALT_RV</th>\n",
       "      <th>ALT_QUAL</th>\n",
       "      <th>ALT_FREQ</th>\n",
       "      <th>TOTAL_DP</th>\n",
       "      <th>PVAL</th>\n",
       "      <th>PASS</th>\n",
       "      <th>GFF_FEATURE</th>\n",
       "      <th>REF_CODON</th>\n",
       "      <th>REF_AA</th>\n",
       "      <th>ALT_CODON</th>\n",
       "      <th>ALT_AA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>9031</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>471</td>\n",
       "      <td>17</td>\n",
       "      <td>48</td>\n",
       "      <td>717</td>\n",
       "      <td>101</td>\n",
       "      <td>39</td>\n",
       "      <td>0.603535</td>\n",
       "      <td>1188</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>TTT</td>\n",
       "      <td>F</td>\n",
       "      <td>TTC</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>9031</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>471</td>\n",
       "      <td>17</td>\n",
       "      <td>48</td>\n",
       "      <td>717</td>\n",
       "      <td>101</td>\n",
       "      <td>39</td>\n",
       "      <td>0.603535</td>\n",
       "      <td>1188</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009725295.1</td>\n",
       "      <td>TTT</td>\n",
       "      <td>F</td>\n",
       "      <td>TTC</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>9038</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>431</td>\n",
       "      <td>22</td>\n",
       "      <td>50</td>\n",
       "      <td>62</td>\n",
       "      <td>2</td>\n",
       "      <td>52</td>\n",
       "      <td>0.125506</td>\n",
       "      <td>494</td>\n",
       "      <td>4.344260e-26</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>GCT</td>\n",
       "      <td>A</td>\n",
       "      <td>TCT</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>9038</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>431</td>\n",
       "      <td>22</td>\n",
       "      <td>50</td>\n",
       "      <td>62</td>\n",
       "      <td>2</td>\n",
       "      <td>52</td>\n",
       "      <td>0.125506</td>\n",
       "      <td>494</td>\n",
       "      <td>4.344260e-26</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009725295.1</td>\n",
       "      <td>GCT</td>\n",
       "      <td>A</td>\n",
       "      <td>TCT</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>9539</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>2685</td>\n",
       "      <td>724</td>\n",
       "      <td>44</td>\n",
       "      <td>124</td>\n",
       "      <td>41</td>\n",
       "      <td>42</td>\n",
       "      <td>0.044097</td>\n",
       "      <td>2812</td>\n",
       "      <td>3.552420e-45</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>CTC</td>\n",
       "      <td>L</td>\n",
       "      <td>TTC</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29215</td>\n",
       "      <td>C</td>\n",
       "      <td>-T</td>\n",
       "      <td>4553</td>\n",
       "      <td>1850</td>\n",
       "      <td>51</td>\n",
       "      <td>266</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>0.037523</td>\n",
       "      <td>7089</td>\n",
       "      <td>8.033940e-49</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29318</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>2844</td>\n",
       "      <td>2272</td>\n",
       "      <td>38</td>\n",
       "      <td>208</td>\n",
       "      <td>189</td>\n",
       "      <td>36</td>\n",
       "      <td>0.068152</td>\n",
       "      <td>3052</td>\n",
       "      <td>6.731180e-64</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724397.2</td>\n",
       "      <td>CAA</td>\n",
       "      <td>Q</td>\n",
       "      <td>TAA</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29323</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>2888</td>\n",
       "      <td>2309</td>\n",
       "      <td>38</td>\n",
       "      <td>135</td>\n",
       "      <td>110</td>\n",
       "      <td>38</td>\n",
       "      <td>0.044584</td>\n",
       "      <td>3028</td>\n",
       "      <td>3.705270e-40</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724397.2</td>\n",
       "      <td>GTC</td>\n",
       "      <td>V</td>\n",
       "      <td>GTT</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29751</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>28</td>\n",
       "      <td>4102</td>\n",
       "      <td>3086</td>\n",
       "      <td>43</td>\n",
       "      <td>0.998054</td>\n",
       "      <td>4110</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29867</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>22</td>\n",
       "      <td>15</td>\n",
       "      <td>35</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>31</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>23</td>\n",
       "      <td>4.509800e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>89 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          REGION    POS REF ALT  REF_DP  REF_RV  REF_QUAL  ALT_DP  ALT_RV  \\\n",
       "36   NC_045512.2   9031   T   C     471      17        48     717     101   \n",
       "37   NC_045512.2   9031   T   C     471      17        48     717     101   \n",
       "38   NC_045512.2   9038   G   T     431      22        50      62       2   \n",
       "39   NC_045512.2   9038   G   T     431      22        50      62       2   \n",
       "40   NC_045512.2   9539   C   T    2685     724        44     124      41   \n",
       "..           ...    ...  ..  ..     ...     ...       ...     ...     ...   \n",
       "120  NC_045512.2  29215   C  -T    4553    1850        51     266       0   \n",
       "121  NC_045512.2  29318   C   T    2844    2272        38     208     189   \n",
       "122  NC_045512.2  29323   C   T    2888    2309        38     135     110   \n",
       "123  NC_045512.2  29751   G   T       1       0        28    4102    3086   \n",
       "124  NC_045512.2  29867   T   A      22      15        35       1       0   \n",
       "\n",
       "     ALT_QUAL  ALT_FREQ  TOTAL_DP          PVAL   PASS         GFF_FEATURE  \\\n",
       "36         39  0.603535      1188  0.000000e+00   True  cds-YP_009724389.1   \n",
       "37         39  0.603535      1188  0.000000e+00   True  cds-YP_009725295.1   \n",
       "38         52  0.125506       494  4.344260e-26   True  cds-YP_009724389.1   \n",
       "39         52  0.125506       494  4.344260e-26   True  cds-YP_009725295.1   \n",
       "40         42  0.044097      2812  3.552420e-45   True  cds-YP_009724389.1   \n",
       "..        ...       ...       ...           ...    ...                 ...   \n",
       "120        20  0.037523      7089  8.033940e-49   True                 NaN   \n",
       "121        36  0.068152      3052  6.731180e-64   True  cds-YP_009724397.2   \n",
       "122        38  0.044584      3028  3.705270e-40   True  cds-YP_009724397.2   \n",
       "123        43  0.998054      4110  0.000000e+00   True                 NaN   \n",
       "124        31  0.043478        23  4.509800e-01  False                 NaN   \n",
       "\n",
       "    REF_CODON REF_AA ALT_CODON ALT_AA  \n",
       "36        TTT      F       TTC      F  \n",
       "37        TTT      F       TTC      F  \n",
       "38        GCT      A       TCT      S  \n",
       "39        GCT      A       TCT      S  \n",
       "40        CTC      L       TTC      F  \n",
       "..        ...    ...       ...    ...  \n",
       "120       NaN    NaN       NaN    NaN  \n",
       "121       CAA      Q       TAA      *  \n",
       "122       GTC      V       GTT      V  \n",
       "123       NaN    NaN       NaN    NaN  \n",
       "124       NaN    NaN       NaN    NaN  \n",
       "\n",
       "[89 rows x 19 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = '/home/gk/analysis/2020.11.16.hCoV19/variants/illumina/SEARCH-4453_L001_L002_L003_L004.tsv'\n",
    "new = pd.read_csv(path, sep='\\t')\n",
    "new.loc[(new['POS'] > 9000)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>REGION</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>REF_DP</th>\n",
       "      <th>REF_RV</th>\n",
       "      <th>REF_QUAL</th>\n",
       "      <th>ALT_DP</th>\n",
       "      <th>ALT_RV</th>\n",
       "      <th>ALT_QUAL</th>\n",
       "      <th>ALT_FREQ</th>\n",
       "      <th>TOTAL_DP</th>\n",
       "      <th>PVAL</th>\n",
       "      <th>PASS</th>\n",
       "      <th>GFF_FEATURE</th>\n",
       "      <th>REF_CODON</th>\n",
       "      <th>REF_AA</th>\n",
       "      <th>ALT_CODON</th>\n",
       "      <th>ALT_AA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>9026</td>\n",
       "      <td>A</td>\n",
       "      <td>-T</td>\n",
       "      <td>607</td>\n",
       "      <td>57</td>\n",
       "      <td>52</td>\n",
       "      <td>32</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>0.035242</td>\n",
       "      <td>908</td>\n",
       "      <td>1.378020e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>9031</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>453</td>\n",
       "      <td>8</td>\n",
       "      <td>49</td>\n",
       "      <td>142</td>\n",
       "      <td>49</td>\n",
       "      <td>57</td>\n",
       "      <td>0.238655</td>\n",
       "      <td>595</td>\n",
       "      <td>5.037440e-63</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>TTT</td>\n",
       "      <td>F</td>\n",
       "      <td>TTC</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>9031</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>453</td>\n",
       "      <td>8</td>\n",
       "      <td>49</td>\n",
       "      <td>142</td>\n",
       "      <td>49</td>\n",
       "      <td>57</td>\n",
       "      <td>0.238655</td>\n",
       "      <td>595</td>\n",
       "      <td>5.037440e-63</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009725295.1</td>\n",
       "      <td>TTT</td>\n",
       "      <td>F</td>\n",
       "      <td>TTC</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>9038</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>416</td>\n",
       "      <td>13</td>\n",
       "      <td>50</td>\n",
       "      <td>62</td>\n",
       "      <td>2</td>\n",
       "      <td>52</td>\n",
       "      <td>0.129436</td>\n",
       "      <td>479</td>\n",
       "      <td>3.060250e-26</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>GCT</td>\n",
       "      <td>A</td>\n",
       "      <td>TCT</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>9038</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>416</td>\n",
       "      <td>13</td>\n",
       "      <td>50</td>\n",
       "      <td>62</td>\n",
       "      <td>2</td>\n",
       "      <td>52</td>\n",
       "      <td>0.129436</td>\n",
       "      <td>479</td>\n",
       "      <td>3.060250e-26</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009725295.1</td>\n",
       "      <td>GCT</td>\n",
       "      <td>A</td>\n",
       "      <td>TCT</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29215</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4164</td>\n",
       "      <td>1701</td>\n",
       "      <td>52</td>\n",
       "      <td>222</td>\n",
       "      <td>102</td>\n",
       "      <td>48</td>\n",
       "      <td>0.050592</td>\n",
       "      <td>4388</td>\n",
       "      <td>2.288810e-91</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724397.2</td>\n",
       "      <td>TTC</td>\n",
       "      <td>F</td>\n",
       "      <td>TTT</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29215</td>\n",
       "      <td>C</td>\n",
       "      <td>-T</td>\n",
       "      <td>4164</td>\n",
       "      <td>1701</td>\n",
       "      <td>52</td>\n",
       "      <td>250</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>0.037476</td>\n",
       "      <td>6671</td>\n",
       "      <td>2.134410e-47</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29318</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>2432</td>\n",
       "      <td>2150</td>\n",
       "      <td>38</td>\n",
       "      <td>194</td>\n",
       "      <td>188</td>\n",
       "      <td>36</td>\n",
       "      <td>0.073877</td>\n",
       "      <td>2626</td>\n",
       "      <td>4.132020e-61</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724397.2</td>\n",
       "      <td>CAA</td>\n",
       "      <td>Q</td>\n",
       "      <td>TAA</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29323</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>2463</td>\n",
       "      <td>2188</td>\n",
       "      <td>38</td>\n",
       "      <td>124</td>\n",
       "      <td>105</td>\n",
       "      <td>38</td>\n",
       "      <td>0.047876</td>\n",
       "      <td>2590</td>\n",
       "      <td>3.694460e-38</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724397.2</td>\n",
       "      <td>GTC</td>\n",
       "      <td>V</td>\n",
       "      <td>GTT</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29751</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>28</td>\n",
       "      <td>3685</td>\n",
       "      <td>2676</td>\n",
       "      <td>43</td>\n",
       "      <td>0.997834</td>\n",
       "      <td>3693</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          REGION    POS REF ALT  REF_DP  REF_RV  REF_QUAL  ALT_DP  ALT_RV  \\\n",
       "50   NC_045512.2   9026   A  -T     607      57        52      32       0   \n",
       "51   NC_045512.2   9031   T   C     453       8        49     142      49   \n",
       "52   NC_045512.2   9031   T   C     453       8        49     142      49   \n",
       "53   NC_045512.2   9038   G   T     416      13        50      62       2   \n",
       "54   NC_045512.2   9038   G   T     416      13        50      62       2   \n",
       "..           ...    ...  ..  ..     ...     ...       ...     ...     ...   \n",
       "137  NC_045512.2  29215   C   T    4164    1701        52     222     102   \n",
       "138  NC_045512.2  29215   C  -T    4164    1701        52     250       0   \n",
       "139  NC_045512.2  29318   C   T    2432    2150        38     194     188   \n",
       "140  NC_045512.2  29323   C   T    2463    2188        38     124     105   \n",
       "141  NC_045512.2  29751   G   T       1       0        28    3685    2676   \n",
       "\n",
       "     ALT_QUAL  ALT_FREQ  TOTAL_DP          PVAL  PASS         GFF_FEATURE  \\\n",
       "50         20  0.035242       908  1.378020e-06  True                 NaN   \n",
       "51         57  0.238655       595  5.037440e-63  True  cds-YP_009724389.1   \n",
       "52         57  0.238655       595  5.037440e-63  True  cds-YP_009725295.1   \n",
       "53         52  0.129436       479  3.060250e-26  True  cds-YP_009724389.1   \n",
       "54         52  0.129436       479  3.060250e-26  True  cds-YP_009725295.1   \n",
       "..        ...       ...       ...           ...   ...                 ...   \n",
       "137        48  0.050592      4388  2.288810e-91  True  cds-YP_009724397.2   \n",
       "138        20  0.037476      6671  2.134410e-47  True                 NaN   \n",
       "139        36  0.073877      2626  4.132020e-61  True  cds-YP_009724397.2   \n",
       "140        38  0.047876      2590  3.694460e-38  True  cds-YP_009724397.2   \n",
       "141        43  0.997834      3693  0.000000e+00  True                 NaN   \n",
       "\n",
       "    REF_CODON REF_AA ALT_CODON ALT_AA  \n",
       "50        NaN    NaN       NaN    NaN  \n",
       "51        TTT      F       TTC      F  \n",
       "52        TTT      F       TTC      F  \n",
       "53        GCT      A       TCT      S  \n",
       "54        GCT      A       TCT      S  \n",
       "..        ...    ...       ...    ...  \n",
       "137       TTC      F       TTT      F  \n",
       "138       NaN    NaN       NaN    NaN  \n",
       "139       CAA      Q       TAA      *  \n",
       "140       GTC      V       GTT      V  \n",
       "141       NaN    NaN       NaN    NaN  \n",
       "\n",
       "[92 rows x 19 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = '/home/al/analysis/11-16.11.2020/variants/illumina/SEARCH-4453_L001_L002_L003_L004.tsv'\n",
    "new = pd.read_csv(path, sep='\\t')\n",
    "new.loc[(new['POS'] > 9000)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>REGION</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>REF_DP</th>\n",
       "      <th>REF_RV</th>\n",
       "      <th>REF_QUAL</th>\n",
       "      <th>ALT_DP</th>\n",
       "      <th>ALT_RV</th>\n",
       "      <th>ALT_QUAL</th>\n",
       "      <th>ALT_FREQ</th>\n",
       "      <th>TOTAL_DP</th>\n",
       "      <th>PVAL</th>\n",
       "      <th>PASS</th>\n",
       "      <th>GFF_FEATURE</th>\n",
       "      <th>REF_CODON</th>\n",
       "      <th>REF_AA</th>\n",
       "      <th>ALT_CODON</th>\n",
       "      <th>ALT_AA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>11074</td>\n",
       "      <td>C</td>\n",
       "      <td>+T</td>\n",
       "      <td>10691</td>\n",
       "      <td>203</td>\n",
       "      <td>38</td>\n",
       "      <td>700</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>0.057148</td>\n",
       "      <td>12249</td>\n",
       "      <td>1.922820e-112</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>11075</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9707</td>\n",
       "      <td>193</td>\n",
       "      <td>38</td>\n",
       "      <td>508</td>\n",
       "      <td>9</td>\n",
       "      <td>36</td>\n",
       "      <td>0.049658</td>\n",
       "      <td>10230</td>\n",
       "      <td>2.117570e-157</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>TTT</td>\n",
       "      <td>F</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>11075</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9707</td>\n",
       "      <td>193</td>\n",
       "      <td>38</td>\n",
       "      <td>508</td>\n",
       "      <td>9</td>\n",
       "      <td>36</td>\n",
       "      <td>0.049658</td>\n",
       "      <td>10230</td>\n",
       "      <td>2.117570e-157</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009725295.1</td>\n",
       "      <td>TTT</td>\n",
       "      <td>F</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>11288</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>18501</td>\n",
       "      <td>11653</td>\n",
       "      <td>47</td>\n",
       "      <td>796</td>\n",
       "      <td>542</td>\n",
       "      <td>46</td>\n",
       "      <td>0.041237</td>\n",
       "      <td>19303</td>\n",
       "      <td>2.814530e-302</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>TCT</td>\n",
       "      <td>S</td>\n",
       "      <td>CCT</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>11288</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>18501</td>\n",
       "      <td>11653</td>\n",
       "      <td>47</td>\n",
       "      <td>796</td>\n",
       "      <td>542</td>\n",
       "      <td>46</td>\n",
       "      <td>0.041237</td>\n",
       "      <td>19303</td>\n",
       "      <td>2.814530e-302</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009725295.1</td>\n",
       "      <td>TCT</td>\n",
       "      <td>S</td>\n",
       "      <td>CCT</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>13135</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>62131</td>\n",
       "      <td>3988</td>\n",
       "      <td>47</td>\n",
       "      <td>2395</td>\n",
       "      <td>86</td>\n",
       "      <td>38</td>\n",
       "      <td>0.037074</td>\n",
       "      <td>64600</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>CCA</td>\n",
       "      <td>P</td>\n",
       "      <td>CCT</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>13135</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>62131</td>\n",
       "      <td>3988</td>\n",
       "      <td>47</td>\n",
       "      <td>2395</td>\n",
       "      <td>86</td>\n",
       "      <td>38</td>\n",
       "      <td>0.037074</td>\n",
       "      <td>64600</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009725295.1</td>\n",
       "      <td>CCA</td>\n",
       "      <td>P</td>\n",
       "      <td>CCT</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>13316</td>\n",
       "      <td>C</td>\n",
       "      <td>-A</td>\n",
       "      <td>61671</td>\n",
       "      <td>43930</td>\n",
       "      <td>45</td>\n",
       "      <td>3204</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>0.039949</td>\n",
       "      <td>80202</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>13333</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>41429</td>\n",
       "      <td>34864</td>\n",
       "      <td>40</td>\n",
       "      <td>1517</td>\n",
       "      <td>1365</td>\n",
       "      <td>38</td>\n",
       "      <td>0.035312</td>\n",
       "      <td>42960</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>TGT</td>\n",
       "      <td>C</td>\n",
       "      <td>TGC</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>13333</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>41429</td>\n",
       "      <td>34864</td>\n",
       "      <td>40</td>\n",
       "      <td>1517</td>\n",
       "      <td>1365</td>\n",
       "      <td>38</td>\n",
       "      <td>0.035312</td>\n",
       "      <td>42960</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009725295.1</td>\n",
       "      <td>TGT</td>\n",
       "      <td>C</td>\n",
       "      <td>TGC</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>14118</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>478</td>\n",
       "      <td>208</td>\n",
       "      <td>49</td>\n",
       "      <td>339</td>\n",
       "      <td>215</td>\n",
       "      <td>46</td>\n",
       "      <td>0.414933</td>\n",
       "      <td>817</td>\n",
       "      <td>1.486840e-152</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>ACG</td>\n",
       "      <td>T</td>\n",
       "      <td>ACA</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>14408</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>15</td>\n",
       "      <td>12</td>\n",
       "      <td>35</td>\n",
       "      <td>44578</td>\n",
       "      <td>11260</td>\n",
       "      <td>50</td>\n",
       "      <td>0.998745</td>\n",
       "      <td>44634</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>CCT</td>\n",
       "      <td>P</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>14952</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>978</td>\n",
       "      <td>45</td>\n",
       "      <td>43</td>\n",
       "      <td>34</td>\n",
       "      <td>2</td>\n",
       "      <td>37</td>\n",
       "      <td>0.033597</td>\n",
       "      <td>1012</td>\n",
       "      <td>1.612410e-11</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>TTT</td>\n",
       "      <td>F</td>\n",
       "      <td>TTC</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>15202</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>50799</td>\n",
       "      <td>3836</td>\n",
       "      <td>36</td>\n",
       "      <td>3213</td>\n",
       "      <td>35</td>\n",
       "      <td>30</td>\n",
       "      <td>0.059475</td>\n",
       "      <td>54023</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>GTA</td>\n",
       "      <td>V</td>\n",
       "      <td>TTA</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>16092</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>28</td>\n",
       "      <td>28</td>\n",
       "      <td>34</td>\n",
       "      <td>36921</td>\n",
       "      <td>27523</td>\n",
       "      <td>44</td>\n",
       "      <td>0.998945</td>\n",
       "      <td>36960</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>TAC</td>\n",
       "      <td>Y</td>\n",
       "      <td>TAT</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>18354</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>16698</td>\n",
       "      <td>724</td>\n",
       "      <td>44</td>\n",
       "      <td>542</td>\n",
       "      <td>25</td>\n",
       "      <td>44</td>\n",
       "      <td>0.031397</td>\n",
       "      <td>17263</td>\n",
       "      <td>2.705210e-192</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>TTA</td>\n",
       "      <td>L</td>\n",
       "      <td>TTC</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>18510</td>\n",
       "      <td>T</td>\n",
       "      <td>-C</td>\n",
       "      <td>21624</td>\n",
       "      <td>10135</td>\n",
       "      <td>49</td>\n",
       "      <td>1151</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>0.038094</td>\n",
       "      <td>30215</td>\n",
       "      <td>2.899270e-180</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>19174</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>13069</td>\n",
       "      <td>6823</td>\n",
       "      <td>51</td>\n",
       "      <td>540</td>\n",
       "      <td>269</td>\n",
       "      <td>51</td>\n",
       "      <td>0.039656</td>\n",
       "      <td>13617</td>\n",
       "      <td>3.998730e-216</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>GAT</td>\n",
       "      <td>D</td>\n",
       "      <td>TAT</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>20268</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>133</td>\n",
       "      <td>10</td>\n",
       "      <td>40</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>133</td>\n",
       "      <td>1.404240e-87</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>TTA</td>\n",
       "      <td>L</td>\n",
       "      <td>TTG</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>21197</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>47</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>28</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>23</td>\n",
       "      <td>4.339620e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "      <td>CGT</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>21208</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>46</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>36</td>\n",
       "      <td>0.038462</td>\n",
       "      <td>26</td>\n",
       "      <td>4.333330e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>ATG</td>\n",
       "      <td>M</td>\n",
       "      <td>TTG</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>21278</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>26</td>\n",
       "      <td>6</td>\n",
       "      <td>51</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>46</td>\n",
       "      <td>0.037037</td>\n",
       "      <td>27</td>\n",
       "      <td>4.090910e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>ATT</td>\n",
       "      <td>I</td>\n",
       "      <td>ACT</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>21427</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>24</td>\n",
       "      <td>12</td>\n",
       "      <td>40</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>36</td>\n",
       "      <td>0.040000</td>\n",
       "      <td>25</td>\n",
       "      <td>4.716980e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>ACT</td>\n",
       "      <td>T</td>\n",
       "      <td>TCT</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>21436</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>42</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>36</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>19</td>\n",
       "      <td>4.634150e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>ATG</td>\n",
       "      <td>M</td>\n",
       "      <td>CTG</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>21466</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>45</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>36</td>\n",
       "      <td>0.041667</td>\n",
       "      <td>24</td>\n",
       "      <td>4.363640e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>ATG</td>\n",
       "      <td>M</td>\n",
       "      <td>CTG</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>21487</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>27</td>\n",
       "      <td>0</td>\n",
       "      <td>47</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>36</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>28</td>\n",
       "      <td>4.117650e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>AAA</td>\n",
       "      <td>K</td>\n",
       "      <td>CAA</td>\n",
       "      <td>Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>21618</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>57</td>\n",
       "      <td>17</td>\n",
       "      <td>58</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>57</td>\n",
       "      <td>2.003130e-43</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724390.1</td>\n",
       "      <td>ACA</td>\n",
       "      <td>T</td>\n",
       "      <td>ATA</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>21954</td>\n",
       "      <td>G</td>\n",
       "      <td>-TGAATT</td>\n",
       "      <td>339</td>\n",
       "      <td>268</td>\n",
       "      <td>41</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>0.046229</td>\n",
       "      <td>411</td>\n",
       "      <td>3.635370e-04</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>23403</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>39</td>\n",
       "      <td>122125</td>\n",
       "      <td>68033</td>\n",
       "      <td>50</td>\n",
       "      <td>0.999035</td>\n",
       "      <td>122243</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724390.1</td>\n",
       "      <td>GAT</td>\n",
       "      <td>D</td>\n",
       "      <td>GGT</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>23696</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>25124</td>\n",
       "      <td>14698</td>\n",
       "      <td>49</td>\n",
       "      <td>1246</td>\n",
       "      <td>800</td>\n",
       "      <td>45</td>\n",
       "      <td>0.047199</td>\n",
       "      <td>26399</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724390.1</td>\n",
       "      <td>ATT</td>\n",
       "      <td>I</td>\n",
       "      <td>TTT</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>25720</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>116</td>\n",
       "      <td>12</td>\n",
       "      <td>45</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>116</td>\n",
       "      <td>4.850830e-79</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724391.1</td>\n",
       "      <td>GCT</td>\n",
       "      <td>A</td>\n",
       "      <td>TCT</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>26325</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>115</td>\n",
       "      <td>2</td>\n",
       "      <td>56</td>\n",
       "      <td>0.991379</td>\n",
       "      <td>116</td>\n",
       "      <td>3.940020e-86</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724392.1</td>\n",
       "      <td>TTG</td>\n",
       "      <td>L</td>\n",
       "      <td>TTA</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>26542</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>27</td>\n",
       "      <td>23</td>\n",
       "      <td>40</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>36</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>28</td>\n",
       "      <td>4.745760e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>cds-YP_009724393.1</td>\n",
       "      <td>ACT</td>\n",
       "      <td>T</td>\n",
       "      <td>AAT</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>27513</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>35</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>36</td>\n",
       "      <td>0.038462</td>\n",
       "      <td>26</td>\n",
       "      <td>5.098040e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>cds-YP_009724395.1</td>\n",
       "      <td>TAC</td>\n",
       "      <td>Y</td>\n",
       "      <td>TAA</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>27527</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>38</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>72</td>\n",
       "      <td>0.033333</td>\n",
       "      <td>30</td>\n",
       "      <td>4.615380e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>cds-YP_009724395.1</td>\n",
       "      <td>CCA</td>\n",
       "      <td>P</td>\n",
       "      <td>CTA</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>27688</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>43</td>\n",
       "      <td>20</td>\n",
       "      <td>45</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>36</td>\n",
       "      <td>0.044444</td>\n",
       "      <td>45</td>\n",
       "      <td>1.884640e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>cds-YP_009724395.1</td>\n",
       "      <td>CCA</td>\n",
       "      <td>P</td>\n",
       "      <td>TCA</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>27753</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>30</td>\n",
       "      <td>24</td>\n",
       "      <td>40</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>36</td>\n",
       "      <td>0.032258</td>\n",
       "      <td>31</td>\n",
       "      <td>4.769230e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>cds-YP_009724395.1</td>\n",
       "      <td>ACA</td>\n",
       "      <td>T</td>\n",
       "      <td>ACC</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>27756</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>31</td>\n",
       "      <td>25</td>\n",
       "      <td>39</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>32</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>32</td>\n",
       "      <td>4.637680e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>cds-YP_009724395.1</td>\n",
       "      <td>GAA</td>\n",
       "      <td>E</td>\n",
       "      <td>GAG</td>\n",
       "      <td>E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>27756</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>31</td>\n",
       "      <td>25</td>\n",
       "      <td>39</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>32</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>32</td>\n",
       "      <td>4.637680e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>cds-YP_009725318.1</td>\n",
       "      <td>ATG</td>\n",
       "      <td>M</td>\n",
       "      <td>GTG</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>27762</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>32</td>\n",
       "      <td>25</td>\n",
       "      <td>42</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>27</td>\n",
       "      <td>0.030303</td>\n",
       "      <td>33</td>\n",
       "      <td>4.583330e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>cds-YP_009725318.1</td>\n",
       "      <td>GAA</td>\n",
       "      <td>E</td>\n",
       "      <td>CAA</td>\n",
       "      <td>Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>27766</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>32</td>\n",
       "      <td>25</td>\n",
       "      <td>41</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "      <td>0.030303</td>\n",
       "      <td>33</td>\n",
       "      <td>4.583330e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>cds-YP_009725318.1</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "      <td>CCT</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>27777</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>31</td>\n",
       "      <td>25</td>\n",
       "      <td>40</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>72</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>32</td>\n",
       "      <td>4.571430e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>cds-YP_009725318.1</td>\n",
       "      <td>GAC</td>\n",
       "      <td>D</td>\n",
       "      <td>TAC</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>27791</td>\n",
       "      <td>C</td>\n",
       "      <td>-T</td>\n",
       "      <td>28</td>\n",
       "      <td>24</td>\n",
       "      <td>39</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>0.032258</td>\n",
       "      <td>31</td>\n",
       "      <td>4.915250e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>27987</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>271</td>\n",
       "      <td>104</td>\n",
       "      <td>52</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>271</td>\n",
       "      <td>1.505160e-199</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724396.1</td>\n",
       "      <td>GTA</td>\n",
       "      <td>V</td>\n",
       "      <td>TTA</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>28214</td>\n",
       "      <td>C</td>\n",
       "      <td>+T</td>\n",
       "      <td>27675</td>\n",
       "      <td>842</td>\n",
       "      <td>43</td>\n",
       "      <td>1071</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>0.030111</td>\n",
       "      <td>35568</td>\n",
       "      <td>3.613220e-123</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>28291</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>24</td>\n",
       "      <td>15</td>\n",
       "      <td>44</td>\n",
       "      <td>28090</td>\n",
       "      <td>12108</td>\n",
       "      <td>50</td>\n",
       "      <td>0.998720</td>\n",
       "      <td>28126</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724397.2</td>\n",
       "      <td>CCC</td>\n",
       "      <td>P</td>\n",
       "      <td>CCT</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>28740</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>7916</td>\n",
       "      <td>4385</td>\n",
       "      <td>39</td>\n",
       "      <td>750</td>\n",
       "      <td>400</td>\n",
       "      <td>38</td>\n",
       "      <td>0.086505</td>\n",
       "      <td>8670</td>\n",
       "      <td>7.963790e-240</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724397.2</td>\n",
       "      <td>GCA</td>\n",
       "      <td>A</td>\n",
       "      <td>GTA</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>28758</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>3386</td>\n",
       "      <td>140</td>\n",
       "      <td>41</td>\n",
       "      <td>365</td>\n",
       "      <td>14</td>\n",
       "      <td>40</td>\n",
       "      <td>0.097152</td>\n",
       "      <td>3757</td>\n",
       "      <td>1.279220e-129</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724397.2</td>\n",
       "      <td>CCT</td>\n",
       "      <td>P</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>28854</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>36</td>\n",
       "      <td>4602</td>\n",
       "      <td>1073</td>\n",
       "      <td>48</td>\n",
       "      <td>0.997615</td>\n",
       "      <td>4613</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724397.2</td>\n",
       "      <td>TCA</td>\n",
       "      <td>S</td>\n",
       "      <td>TTA</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29825</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>32</td>\n",
       "      <td>5262</td>\n",
       "      <td>4163</td>\n",
       "      <td>42</td>\n",
       "      <td>0.998861</td>\n",
       "      <td>5268</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         REGION    POS REF      ALT  REF_DP  REF_RV  REF_QUAL  ALT_DP  ALT_RV  \\\n",
       "40  NC_045512.2  11074   C       +T   10691     203        38     700       0   \n",
       "41  NC_045512.2  11075   T        C    9707     193        38     508       9   \n",
       "42  NC_045512.2  11075   T        C    9707     193        38     508       9   \n",
       "43  NC_045512.2  11288   T        C   18501   11653        47     796     542   \n",
       "44  NC_045512.2  11288   T        C   18501   11653        47     796     542   \n",
       "45  NC_045512.2  13135   A        T   62131    3988        47    2395      86   \n",
       "46  NC_045512.2  13135   A        T   62131    3988        47    2395      86   \n",
       "47  NC_045512.2  13316   C       -A   61671   43930        45    3204       0   \n",
       "48  NC_045512.2  13333   T        C   41429   34864        40    1517    1365   \n",
       "49  NC_045512.2  13333   T        C   41429   34864        40    1517    1365   \n",
       "50  NC_045512.2  14118   G        A     478     208        49     339     215   \n",
       "51  NC_045512.2  14408   C        T      15      12        35   44578   11260   \n",
       "52  NC_045512.2  14952   T        C     978      45        43      34       2   \n",
       "53  NC_045512.2  15202   G        T   50799    3836        36    3213      35   \n",
       "54  NC_045512.2  16092   C        T      28      28        34   36921   27523   \n",
       "55  NC_045512.2  18354   A        C   16698     724        44     542      25   \n",
       "56  NC_045512.2  18510   T       -C   21624   10135        49    1151       0   \n",
       "57  NC_045512.2  19174   G        T   13069    6823        51     540     269   \n",
       "58  NC_045512.2  20268   A        G       0       0         0     133      10   \n",
       "59  NC_045512.2  21197   T        G      22       0        47       1       0   \n",
       "60  NC_045512.2  21208   A        T      25       0        46       1       0   \n",
       "61  NC_045512.2  21278   T        C      26       6        51       1       0   \n",
       "62  NC_045512.2  21427   A        T      24      12        40       1       1   \n",
       "63  NC_045512.2  21436   A        C      18       0        42       1       1   \n",
       "64  NC_045512.2  21466   A        C      23       0        45       1       1   \n",
       "65  NC_045512.2  21487   A        C      27       0        47       1       1   \n",
       "66  NC_045512.2  21618   C        T       0       0         0      57      17   \n",
       "67  NC_045512.2  21954   G  -TGAATT     339     268        41      19       0   \n",
       "68  NC_045512.2  23403   A        G      10       5        39  122125   68033   \n",
       "69  NC_045512.2  23696   A        T   25124   14698        49    1246     800   \n",
       "70  NC_045512.2  25720   G        T       0       0         0     116      12   \n",
       "71  NC_045512.2  26325   G        A       0       0         0     115       2   \n",
       "72  NC_045512.2  26542   C        A      27      23        40       1       1   \n",
       "73  NC_045512.2  27513   C        A      25       0        35       1       0   \n",
       "74  NC_045512.2  27527   C        T      29       0        38       1       0   \n",
       "75  NC_045512.2  27688   C        T      43      20        45       2       2   \n",
       "76  NC_045512.2  27753   A        C      30      24        40       1       1   \n",
       "77  NC_045512.2  27756   A        G      31      25        39       1       1   \n",
       "78  NC_045512.2  27756   A        G      31      25        39       1       1   \n",
       "79  NC_045512.2  27762   G        C      32      25        42       1       1   \n",
       "80  NC_045512.2  27766   T        C      32      25        41       1       1   \n",
       "81  NC_045512.2  27777   G        T      31      25        40       1       0   \n",
       "82  NC_045512.2  27791   C       -T      28      24        39       1       0   \n",
       "83  NC_045512.2  27987   G        T       0       0         0     271     104   \n",
       "84  NC_045512.2  28214   C       +T   27675     842        43    1071       0   \n",
       "85  NC_045512.2  28291   C        T      24      15        44   28090   12108   \n",
       "86  NC_045512.2  28740   C        T    7916    4385        39     750     400   \n",
       "87  NC_045512.2  28758   C        T    3386     140        41     365      14   \n",
       "88  NC_045512.2  28854   C        T       1       1        36    4602    1073   \n",
       "89  NC_045512.2  29825   G        T       2       1        32    5262    4163   \n",
       "\n",
       "    ALT_QUAL  ALT_FREQ  TOTAL_DP           PVAL   PASS         GFF_FEATURE  \\\n",
       "40        20  0.057148     12249  1.922820e-112   True                 NaN   \n",
       "41        36  0.049658     10230  2.117570e-157   True  cds-YP_009724389.1   \n",
       "42        36  0.049658     10230  2.117570e-157   True  cds-YP_009725295.1   \n",
       "43        46  0.041237     19303  2.814530e-302   True  cds-YP_009724389.1   \n",
       "44        46  0.041237     19303  2.814530e-302   True  cds-YP_009725295.1   \n",
       "45        38  0.037074     64600   0.000000e+00   True  cds-YP_009724389.1   \n",
       "46        38  0.037074     64600   0.000000e+00   True  cds-YP_009725295.1   \n",
       "47        20  0.039949     80202   0.000000e+00   True                 NaN   \n",
       "48        38  0.035312     42960   0.000000e+00   True  cds-YP_009724389.1   \n",
       "49        38  0.035312     42960   0.000000e+00   True  cds-YP_009725295.1   \n",
       "50        46  0.414933       817  1.486840e-152   True  cds-YP_009724389.1   \n",
       "51        50  0.998745     44634   0.000000e+00   True  cds-YP_009724389.1   \n",
       "52        37  0.033597      1012   1.612410e-11   True  cds-YP_009724389.1   \n",
       "53        30  0.059475     54023   0.000000e+00   True  cds-YP_009724389.1   \n",
       "54        44  0.998945     36960   0.000000e+00   True  cds-YP_009724389.1   \n",
       "55        44  0.031397     17263  2.705210e-192   True  cds-YP_009724389.1   \n",
       "56        20  0.038094     30215  2.899270e-180   True                 NaN   \n",
       "57        51  0.039656     13617  3.998730e-216   True  cds-YP_009724389.1   \n",
       "58        40  1.000000       133   1.404240e-87   True  cds-YP_009724389.1   \n",
       "59        28  0.043478        23   4.339620e-01  False  cds-YP_009724389.1   \n",
       "60        36  0.038462        26   4.333330e-01  False  cds-YP_009724389.1   \n",
       "61        46  0.037037        27   4.090910e-01  False  cds-YP_009724389.1   \n",
       "62        36  0.040000        25   4.716980e-01  False  cds-YP_009724389.1   \n",
       "63        36  0.052632        19   4.634150e-01  False  cds-YP_009724389.1   \n",
       "64        36  0.041667        24   4.363640e-01  False  cds-YP_009724389.1   \n",
       "65        36  0.035714        28   4.117650e-01  False  cds-YP_009724389.1   \n",
       "66        58  1.000000        57   2.003130e-43   True  cds-YP_009724390.1   \n",
       "67        20  0.046229       411   3.635370e-04   True                 NaN   \n",
       "68        50  0.999035    122243   0.000000e+00   True  cds-YP_009724390.1   \n",
       "69        45  0.047199     26399   0.000000e+00   True  cds-YP_009724390.1   \n",
       "70        45  1.000000       116   4.850830e-79   True  cds-YP_009724391.1   \n",
       "71        56  0.991379       116   3.940020e-86   True  cds-YP_009724392.1   \n",
       "72        36  0.035714        28   4.745760e-01  False  cds-YP_009724393.1   \n",
       "73        36  0.038462        26   5.098040e-01  False  cds-YP_009724395.1   \n",
       "74        72  0.033333        30   4.615380e-01  False  cds-YP_009724395.1   \n",
       "75        36  0.044444        45   1.884640e-01  False  cds-YP_009724395.1   \n",
       "76        36  0.032258        31   4.769230e-01  False  cds-YP_009724395.1   \n",
       "77        32  0.031250        32   4.637680e-01  False  cds-YP_009724395.1   \n",
       "78        32  0.031250        32   4.637680e-01  False  cds-YP_009725318.1   \n",
       "79        27  0.030303        33   4.583330e-01  False  cds-YP_009725318.1   \n",
       "80        21  0.030303        33   4.583330e-01  False  cds-YP_009725318.1   \n",
       "81        72  0.031250        32   4.571430e-01  False  cds-YP_009725318.1   \n",
       "82        20  0.032258        31   4.915250e-01  False                 NaN   \n",
       "83        52  1.000000       271  1.505160e-199   True  cds-YP_009724396.1   \n",
       "84        20  0.030111     35568  3.613220e-123   True                 NaN   \n",
       "85        50  0.998720     28126   0.000000e+00   True  cds-YP_009724397.2   \n",
       "86        38  0.086505      8670  7.963790e-240   True  cds-YP_009724397.2   \n",
       "87        40  0.097152      3757  1.279220e-129   True  cds-YP_009724397.2   \n",
       "88        48  0.997615      4613   0.000000e+00   True  cds-YP_009724397.2   \n",
       "89        42  0.998861      5268   0.000000e+00   True                 NaN   \n",
       "\n",
       "   REF_CODON REF_AA ALT_CODON ALT_AA  \n",
       "40       NaN    NaN       NaN    NaN  \n",
       "41       TTT      F       CTT      L  \n",
       "42       TTT      F       CTT      L  \n",
       "43       TCT      S       CCT      P  \n",
       "44       TCT      S       CCT      P  \n",
       "45       CCA      P       CCT      P  \n",
       "46       CCA      P       CCT      P  \n",
       "47       NaN    NaN       NaN    NaN  \n",
       "48       TGT      C       TGC      C  \n",
       "49       TGT      C       TGC      C  \n",
       "50       ACG      T       ACA      T  \n",
       "51       CCT      P       CTT      L  \n",
       "52       TTT      F       TTC      F  \n",
       "53       GTA      V       TTA      L  \n",
       "54       TAC      Y       TAT      Y  \n",
       "55       TTA      L       TTC      F  \n",
       "56       NaN    NaN       NaN    NaN  \n",
       "57       GAT      D       TAT      Y  \n",
       "58       TTA      L       TTG      L  \n",
       "59       CTT      L       CGT      R  \n",
       "60       ATG      M       TTG      L  \n",
       "61       ATT      I       ACT      T  \n",
       "62       ACT      T       TCT      S  \n",
       "63       ATG      M       CTG      L  \n",
       "64       ATG      M       CTG      L  \n",
       "65       AAA      K       CAA      Q  \n",
       "66       ACA      T       ATA      I  \n",
       "67       NaN    NaN       NaN    NaN  \n",
       "68       GAT      D       GGT      G  \n",
       "69       ATT      I       TTT      F  \n",
       "70       GCT      A       TCT      S  \n",
       "71       TTG      L       TTA      L  \n",
       "72       ACT      T       AAT      N  \n",
       "73       TAC      Y       TAA      *  \n",
       "74       CCA      P       CTA      L  \n",
       "75       CCA      P       TCA      S  \n",
       "76       ACA      T       ACC      T  \n",
       "77       GAA      E       GAG      E  \n",
       "78       ATG      M       GTG      V  \n",
       "79       GAA      E       CAA      Q  \n",
       "80       CTT      L       CCT      P  \n",
       "81       GAC      D       TAC      Y  \n",
       "82       NaN    NaN       NaN    NaN  \n",
       "83       GTA      V       TTA      L  \n",
       "84       NaN    NaN       NaN    NaN  \n",
       "85       CCC      P       CCT      P  \n",
       "86       GCA      A       GTA      V  \n",
       "87       CCT      P       CTT      L  \n",
       "88       TCA      S       TTA      L  \n",
       "89       NaN    NaN       NaN    NaN  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = '/home/al/analysis/11-16.11.2020/variants/illumina/SEARCH-4459_L001_L002_L003_L004.tsv'\n",
    "new = pd.read_csv(path, sep='\\t')\n",
    "new.loc[(new['POS'] > 9000)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>REGION</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>REF_DP</th>\n",
       "      <th>REF_RV</th>\n",
       "      <th>REF_QUAL</th>\n",
       "      <th>ALT_DP</th>\n",
       "      <th>ALT_RV</th>\n",
       "      <th>ALT_QUAL</th>\n",
       "      <th>ALT_FREQ</th>\n",
       "      <th>TOTAL_DP</th>\n",
       "      <th>PVAL</th>\n",
       "      <th>PASS</th>\n",
       "      <th>GFF_FEATURE</th>\n",
       "      <th>REF_CODON</th>\n",
       "      <th>REF_AA</th>\n",
       "      <th>ALT_CODON</th>\n",
       "      <th>ALT_AA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>14118</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>561</td>\n",
       "      <td>267</td>\n",
       "      <td>47</td>\n",
       "      <td>626</td>\n",
       "      <td>290</td>\n",
       "      <td>40</td>\n",
       "      <td>0.527380</td>\n",
       "      <td>1187</td>\n",
       "      <td>8.632320e-283</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>ACG</td>\n",
       "      <td>T</td>\n",
       "      <td>ACA</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>14408</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>34</td>\n",
       "      <td>7342</td>\n",
       "      <td>3603</td>\n",
       "      <td>37</td>\n",
       "      <td>0.998912</td>\n",
       "      <td>7350</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>CCT</td>\n",
       "      <td>P</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>16092</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>31</td>\n",
       "      <td>6663</td>\n",
       "      <td>4069</td>\n",
       "      <td>38</td>\n",
       "      <td>0.999400</td>\n",
       "      <td>6667</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>TAC</td>\n",
       "      <td>Y</td>\n",
       "      <td>TAT</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>18354</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>6970</td>\n",
       "      <td>834</td>\n",
       "      <td>38</td>\n",
       "      <td>239</td>\n",
       "      <td>34</td>\n",
       "      <td>39</td>\n",
       "      <td>0.033135</td>\n",
       "      <td>7213</td>\n",
       "      <td>7.088820e-68</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>TTA</td>\n",
       "      <td>L</td>\n",
       "      <td>TTC</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>18510</td>\n",
       "      <td>T</td>\n",
       "      <td>-C</td>\n",
       "      <td>7163</td>\n",
       "      <td>3900</td>\n",
       "      <td>39</td>\n",
       "      <td>339</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>0.042396</td>\n",
       "      <td>7996</td>\n",
       "      <td>1.636180e-44</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>19174</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>6397</td>\n",
       "      <td>3693</td>\n",
       "      <td>42</td>\n",
       "      <td>242</td>\n",
       "      <td>126</td>\n",
       "      <td>43</td>\n",
       "      <td>0.036424</td>\n",
       "      <td>6644</td>\n",
       "      <td>7.052330e-85</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>GAT</td>\n",
       "      <td>D</td>\n",
       "      <td>TAT</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>20268</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>360</td>\n",
       "      <td>182</td>\n",
       "      <td>39</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>360</td>\n",
       "      <td>1.027110e-232</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>TTA</td>\n",
       "      <td>L</td>\n",
       "      <td>TTG</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>21618</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>127</td>\n",
       "      <td>27</td>\n",
       "      <td>46</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>127</td>\n",
       "      <td>4.117970e-89</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724390.1</td>\n",
       "      <td>ACA</td>\n",
       "      <td>T</td>\n",
       "      <td>ATA</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>21954</td>\n",
       "      <td>G</td>\n",
       "      <td>-TGAATT</td>\n",
       "      <td>406</td>\n",
       "      <td>307</td>\n",
       "      <td>41</td>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>0.047131</td>\n",
       "      <td>488</td>\n",
       "      <td>3.421190e-05</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>23403</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>36</td>\n",
       "      <td>7492</td>\n",
       "      <td>3476</td>\n",
       "      <td>36</td>\n",
       "      <td>0.999067</td>\n",
       "      <td>7499</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724390.1</td>\n",
       "      <td>GAT</td>\n",
       "      <td>D</td>\n",
       "      <td>GGT</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>23696</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>6477</td>\n",
       "      <td>3557</td>\n",
       "      <td>39</td>\n",
       "      <td>252</td>\n",
       "      <td>144</td>\n",
       "      <td>37</td>\n",
       "      <td>0.037417</td>\n",
       "      <td>6735</td>\n",
       "      <td>3.247320e-75</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724390.1</td>\n",
       "      <td>ATT</td>\n",
       "      <td>I</td>\n",
       "      <td>TTT</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>25720</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>145</td>\n",
       "      <td>45</td>\n",
       "      <td>45</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>145</td>\n",
       "      <td>2.182030e-99</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724391.1</td>\n",
       "      <td>GCT</td>\n",
       "      <td>A</td>\n",
       "      <td>TCT</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>26325</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>248</td>\n",
       "      <td>124</td>\n",
       "      <td>45</td>\n",
       "      <td>0.992000</td>\n",
       "      <td>250</td>\n",
       "      <td>1.033070e-170</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724392.1</td>\n",
       "      <td>TTG</td>\n",
       "      <td>L</td>\n",
       "      <td>TTA</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>27688</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>73</td>\n",
       "      <td>29</td>\n",
       "      <td>40</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>36</td>\n",
       "      <td>0.063291</td>\n",
       "      <td>79</td>\n",
       "      <td>1.784520e-02</td>\n",
       "      <td>False</td>\n",
       "      <td>cds-YP_009724395.1</td>\n",
       "      <td>CCA</td>\n",
       "      <td>P</td>\n",
       "      <td>TCA</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>27805</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>26</td>\n",
       "      <td>18</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>36</td>\n",
       "      <td>0.037037</td>\n",
       "      <td>27</td>\n",
       "      <td>4.909090e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>cds-YP_009725318.1</td>\n",
       "      <td>CTG</td>\n",
       "      <td>L</td>\n",
       "      <td>CAG</td>\n",
       "      <td>Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>27828</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>132</td>\n",
       "      <td>15</td>\n",
       "      <td>39</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>37</td>\n",
       "      <td>0.089655</td>\n",
       "      <td>145</td>\n",
       "      <td>3.803790e-05</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009725318.1</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "      <td>TTT</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>27987</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>309</td>\n",
       "      <td>120</td>\n",
       "      <td>50</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>309</td>\n",
       "      <td>1.554880e-223</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724396.1</td>\n",
       "      <td>GTA</td>\n",
       "      <td>V</td>\n",
       "      <td>TTA</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>28214</td>\n",
       "      <td>C</td>\n",
       "      <td>+T</td>\n",
       "      <td>7084</td>\n",
       "      <td>642</td>\n",
       "      <td>37</td>\n",
       "      <td>265</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>0.033204</td>\n",
       "      <td>7981</td>\n",
       "      <td>5.761950e-29</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>28291</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>30</td>\n",
       "      <td>7062</td>\n",
       "      <td>4185</td>\n",
       "      <td>38</td>\n",
       "      <td>0.997739</td>\n",
       "      <td>7078</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724397.2</td>\n",
       "      <td>CCC</td>\n",
       "      <td>P</td>\n",
       "      <td>CCT</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>28740</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>5918</td>\n",
       "      <td>1987</td>\n",
       "      <td>37</td>\n",
       "      <td>836</td>\n",
       "      <td>265</td>\n",
       "      <td>37</td>\n",
       "      <td>0.123687</td>\n",
       "      <td>6759</td>\n",
       "      <td>2.498490e-269</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724397.2</td>\n",
       "      <td>GCA</td>\n",
       "      <td>A</td>\n",
       "      <td>GTA</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>28758</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>6005</td>\n",
       "      <td>1726</td>\n",
       "      <td>37</td>\n",
       "      <td>908</td>\n",
       "      <td>262</td>\n",
       "      <td>37</td>\n",
       "      <td>0.131176</td>\n",
       "      <td>6922</td>\n",
       "      <td>5.490110e-293</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724397.2</td>\n",
       "      <td>CCT</td>\n",
       "      <td>P</td>\n",
       "      <td>CTT</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>28854</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>36</td>\n",
       "      <td>7202</td>\n",
       "      <td>2198</td>\n",
       "      <td>38</td>\n",
       "      <td>0.998890</td>\n",
       "      <td>7210</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724397.2</td>\n",
       "      <td>TCA</td>\n",
       "      <td>S</td>\n",
       "      <td>TTA</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29051</td>\n",
       "      <td>C</td>\n",
       "      <td>+A</td>\n",
       "      <td>7316</td>\n",
       "      <td>1929</td>\n",
       "      <td>37</td>\n",
       "      <td>263</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>0.032883</td>\n",
       "      <td>7998</td>\n",
       "      <td>2.792320e-27</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29308</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>6225</td>\n",
       "      <td>2735</td>\n",
       "      <td>36</td>\n",
       "      <td>242</td>\n",
       "      <td>92</td>\n",
       "      <td>36</td>\n",
       "      <td>0.037380</td>\n",
       "      <td>6474</td>\n",
       "      <td>4.885100e-68</td>\n",
       "      <td>True</td>\n",
       "      <td>cds-YP_009724397.2</td>\n",
       "      <td>AAT</td>\n",
       "      <td>N</td>\n",
       "      <td>AAA</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>29825</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>28</td>\n",
       "      <td>1953</td>\n",
       "      <td>1343</td>\n",
       "      <td>43</td>\n",
       "      <td>0.997956</td>\n",
       "      <td>1957</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         REGION    POS REF      ALT  REF_DP  REF_RV  REF_QUAL  ALT_DP  ALT_RV  \\\n",
       "43  NC_045512.2  14118   G        A     561     267        47     626     290   \n",
       "44  NC_045512.2  14408   C        T       3       3        34    7342    3603   \n",
       "45  NC_045512.2  16092   C        T       2       2        31    6663    4069   \n",
       "46  NC_045512.2  18354   A        C    6970     834        38     239      34   \n",
       "47  NC_045512.2  18510   T       -C    7163    3900        39     339       0   \n",
       "48  NC_045512.2  19174   G        T    6397    3693        42     242     126   \n",
       "49  NC_045512.2  20268   A        G       0       0         0     360     182   \n",
       "50  NC_045512.2  21618   C        T       0       0         0     127      27   \n",
       "51  NC_045512.2  21954   G  -TGAATT     406     307        41      23       0   \n",
       "52  NC_045512.2  23403   A        G       1       1        36    7492    3476   \n",
       "53  NC_045512.2  23696   A        T    6477    3557        39     252     144   \n",
       "54  NC_045512.2  25720   G        T       0       0         0     145      45   \n",
       "55  NC_045512.2  26325   G        A       0       0         0     248     124   \n",
       "56  NC_045512.2  27688   C        T      73      29        40       5       2   \n",
       "57  NC_045512.2  27805   T        A      26      18        37       1       0   \n",
       "58  NC_045512.2  27828   C        T     132      15        39      13       1   \n",
       "59  NC_045512.2  27987   G        T       0       0         0     309     120   \n",
       "60  NC_045512.2  28214   C       +T    7084     642        37     265       0   \n",
       "61  NC_045512.2  28291   C        T       7       6        30    7062    4185   \n",
       "62  NC_045512.2  28740   C        T    5918    1987        37     836     265   \n",
       "63  NC_045512.2  28758   C        T    6005    1726        37     908     262   \n",
       "64  NC_045512.2  28854   C        T       1       1        36    7202    2198   \n",
       "65  NC_045512.2  29051   C       +A    7316    1929        37     263       0   \n",
       "66  NC_045512.2  29308   T        A    6225    2735        36     242      92   \n",
       "67  NC_045512.2  29825   G        T       1       0        28    1953    1343   \n",
       "\n",
       "    ALT_QUAL  ALT_FREQ  TOTAL_DP           PVAL   PASS         GFF_FEATURE  \\\n",
       "43        40  0.527380      1187  8.632320e-283   True  cds-YP_009724389.1   \n",
       "44        37  0.998912      7350   0.000000e+00   True  cds-YP_009724389.1   \n",
       "45        38  0.999400      6667   0.000000e+00   True  cds-YP_009724389.1   \n",
       "46        39  0.033135      7213   7.088820e-68   True  cds-YP_009724389.1   \n",
       "47        20  0.042396      7996   1.636180e-44   True                 NaN   \n",
       "48        43  0.036424      6644   7.052330e-85   True  cds-YP_009724389.1   \n",
       "49        39  1.000000       360  1.027110e-232   True  cds-YP_009724389.1   \n",
       "50        46  1.000000       127   4.117970e-89   True  cds-YP_009724390.1   \n",
       "51        20  0.047131       488   3.421190e-05   True                 NaN   \n",
       "52        36  0.999067      7499   0.000000e+00   True  cds-YP_009724390.1   \n",
       "53        37  0.037417      6735   3.247320e-75   True  cds-YP_009724390.1   \n",
       "54        45  1.000000       145   2.182030e-99   True  cds-YP_009724391.1   \n",
       "55        45  0.992000       250  1.033070e-170   True  cds-YP_009724392.1   \n",
       "56        36  0.063291        79   1.784520e-02  False  cds-YP_009724395.1   \n",
       "57        36  0.037037        27   4.909090e-01  False  cds-YP_009725318.1   \n",
       "58        37  0.089655       145   3.803790e-05   True  cds-YP_009725318.1   \n",
       "59        50  1.000000       309  1.554880e-223   True  cds-YP_009724396.1   \n",
       "60        20  0.033204      7981   5.761950e-29   True                 NaN   \n",
       "61        38  0.997739      7078   0.000000e+00   True  cds-YP_009724397.2   \n",
       "62        37  0.123687      6759  2.498490e-269   True  cds-YP_009724397.2   \n",
       "63        37  0.131176      6922  5.490110e-293   True  cds-YP_009724397.2   \n",
       "64        38  0.998890      7210   0.000000e+00   True  cds-YP_009724397.2   \n",
       "65        20  0.032883      7998   2.792320e-27   True                 NaN   \n",
       "66        36  0.037380      6474   4.885100e-68   True  cds-YP_009724397.2   \n",
       "67        43  0.997956      1957   0.000000e+00   True                 NaN   \n",
       "\n",
       "   REF_CODON REF_AA ALT_CODON ALT_AA  \n",
       "43       ACG      T       ACA      T  \n",
       "44       CCT      P       CTT      L  \n",
       "45       TAC      Y       TAT      Y  \n",
       "46       TTA      L       TTC      F  \n",
       "47       NaN    NaN       NaN    NaN  \n",
       "48       GAT      D       TAT      Y  \n",
       "49       TTA      L       TTG      L  \n",
       "50       ACA      T       ATA      I  \n",
       "51       NaN    NaN       NaN    NaN  \n",
       "52       GAT      D       GGT      G  \n",
       "53       ATT      I       TTT      F  \n",
       "54       GCT      A       TCT      S  \n",
       "55       TTG      L       TTA      L  \n",
       "56       CCA      P       TCA      S  \n",
       "57       CTG      L       CAG      Q  \n",
       "58       CTT      L       TTT      F  \n",
       "59       GTA      V       TTA      L  \n",
       "60       NaN    NaN       NaN    NaN  \n",
       "61       CCC      P       CCT      P  \n",
       "62       GCA      A       GTA      V  \n",
       "63       CCT      P       CTT      L  \n",
       "64       TCA      S       TTA      L  \n",
       "65       NaN    NaN       NaN    NaN  \n",
       "66       AAT      N       AAA      K  \n",
       "67       NaN    NaN       NaN    NaN  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = '/home/gk/analysis/2020.11.16.hCoV19/variants/illumina/SEARCH-4459_L001_L002_L003_L004.tsv'\n",
    "df = pd.read_csv(path, sep='\\t')\n",
    "df.loc[(df['POS'] > 14000)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "243"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "analysis_pth = '/home/al/analysis/11-16.11.2020/consensus_sequences/illumina'\n",
    "new_cns_filepaths = get_filepaths(sample_ids=old_cns_filepaths.keys(),\n",
    "                                  analysis_path=analysis_pth,\n",
    "                                  data_fmt='fa')\n",
    "len(new_cns_filepaths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "243"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get cns seqs\n",
    "new_cns_filepaths = {s_id: fp for s_id, fp in new_cns_filepaths.items() if s_id in old_cns_filepaths.keys()}\n",
    "len(new_cns_filepaths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4130-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4130-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4189-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4189-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4220-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4220-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4192-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4192-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4184-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4184-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4134-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4134-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4099-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4099-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4390-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4390-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4347-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4347-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4195-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4195-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4205-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4205-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4165-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4165-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4223-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4223-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4265-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4265-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4395-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4395-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4218-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4218-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4374-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4374-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4243-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4243-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4479-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4479-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4149-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4149-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4429-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4429-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4412-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4412-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4471-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4471-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4459-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4459-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4123-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4123-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4294-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4294-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4331-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4331-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4303-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4303-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4108-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4108-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4263-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4263-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4481-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4481-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4233-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4233-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4148-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4148-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4310-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4310-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4369-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4369-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4129-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4129-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4411-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4411-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4427-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4427-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4428-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4428-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4267-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4267-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4102-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4102-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4478-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4478-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4376-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4376-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4196-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4196-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4420-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4420-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4309-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4309-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4143-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4143-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4269-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4269-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4464-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4464-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4236-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4236-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4365-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4365-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4106-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4106-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4325-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4325-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4484-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4484-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4417-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4417-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4295-SBD.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4295-SBD.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4289-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4289-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4407-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4407-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4185-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4185-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4449-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4449-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4431-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4431-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4142-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4142-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4474-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4474-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4103-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4103-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4314-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4314-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4329-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4329-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4356-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4356-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4344-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4344-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4447-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4447-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4144-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4144-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4270-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4270-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4181-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4181-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4139-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4139-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4166-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4166-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4191-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4191-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4320-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4320-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4171-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4171-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4274-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4274-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4340-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4340-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4463-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4463-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4381-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4381-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4351-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4351-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4371-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4371-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4264-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4264-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4346-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4346-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4384-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4384-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4409-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4409-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4328-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4328-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4388-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4388-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4453-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4453-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4224-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4224-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4279-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4279-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4235-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4235-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4197-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4197-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4296-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4296-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4122-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4122-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4182-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4182-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4330-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4330-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4298-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4298-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4140-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4140-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4462-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4462-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4437-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4437-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4311-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4311-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4247-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4247-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4131-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4131-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4415-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4415-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4187-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4187-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4404-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4404-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4401-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4401-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4126-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4126-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4307-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4307-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4251-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4251-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4432-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4432-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4146-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4146-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4475-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4475-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4245-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4245-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4287-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4287-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4109-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4109-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4450-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4450-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4111-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4111-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4458-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4458-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4282-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4282-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4413-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4413-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4380-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4380-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4391-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4391-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4155-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4155-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4256-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4256-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4297-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4297-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4436-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4436-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4199-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4199-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4364-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4364-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4249-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4249-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4305-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4305-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4164-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4164-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4306-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4306-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4259-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4259-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4138-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4138-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4435-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4435-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4398-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4398-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4260-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4260-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4156-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4156-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4237-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4237-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4317-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4317-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4332-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4332-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4186-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4186-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4116-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4116-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4211-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4211-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4176-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4176-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4288-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4288-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4221-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4221-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4177-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4177-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4190-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4190-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4336-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4336-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4337-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4337-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4183-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4183-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4399-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4399-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4229-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4229-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4343-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4343-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4262-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4262-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4125-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4125-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4276-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4276-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4292-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4292-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4280-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4280-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4358-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4358-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4286-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4286-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4461-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4461-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4124-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4124-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4253-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4253-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4145-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4145-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4238-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4238-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4290-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4290-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4230-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4230-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4455-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4455-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4322-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4322-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4104-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4104-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4385-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4385-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4468-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4468-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4304-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4304-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4326-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4326-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4451-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4451-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4242-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4242-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4157-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4157-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4355-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4355-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4141-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4141-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4152-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4152-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4284-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4284-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4452-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4452-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4440-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4440-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4266-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4266-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4472-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4472-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4448-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4448-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4308-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4308-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4405-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4405-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4283-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4283-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4128-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4128-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4442-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4442-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4252-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4252-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4433-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4433-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4419-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4419-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4316-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4316-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4273-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4273-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4319-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4319-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4261-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4261-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4430-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4430-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4172-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4172-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4254-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4254-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4366-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4366-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4150-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4150-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4248-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4248-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4255-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4255-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4315-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4315-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4439-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4439-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4342-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4342-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4225-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4225-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4466-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4466-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4213-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4213-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4313-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4313-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4173-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4173-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4387-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4387-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4302-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4302-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4318-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4318-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4188-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4188-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4250-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4250-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4206-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4206-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4107-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4107-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4291-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4291-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4467-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4467-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4217-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4217-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4359-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4359-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4133-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4133-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4425-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4425-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4293-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4293-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4209-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4209-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4473-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4473-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4400-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4400-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4194-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4194-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4216-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4216-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4208-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4208-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4203-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4203-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4151-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4151-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4285-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4285-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4389-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4389-SAN.fasta\n",
      "Transferring /home/al/code/HCoV-19-Genomics/consensus_sequences/SEARCH-4300-SAN.fasta to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4300-SAN.fasta\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def rename_fasta_hdrs(out, cns_fps: dict, hdr_suffix: str):\n",
    "    for s_id, cons_fp in cns_fps.items():\n",
    "        n = os.path.join(out, os.path.basename(cons_fp))\n",
    "        print(\"Transferring {} to {}\".format(cons_fp, n))\n",
    "        rec = SeqIO.read(cons_fp, format='fasta')\n",
    "        hdr = f\"{s_id}{hdr_suffix}\"\n",
    "        rec.name = hdr\n",
    "        rec.id = hdr\n",
    "        rec.description = hdr\n",
    "        with open(n, \"w\") as fout:\n",
    "            SeqIO.write(rec, fout, format=\"fasta\")\n",
    "            fout.close()\n",
    "    return 0\n",
    "\n",
    "out_dir = \"/home/al/analysis/ivar_retrospect/consensus/\"\n",
    "rename_fasta_hdrs(out_dir, old_cns_filepaths, hdr_suffix='_old')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4130_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4130_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4189_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4189_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4220_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4220_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4192_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4192_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4184_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4184_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4134_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4134_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4099_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4099_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4390_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4390_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4347_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4347_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4195_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4195_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4205_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4205_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4165_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4165_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4223_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4223_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4265_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4265_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4395_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4395_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4218_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4218_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4374_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4374_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4243_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4243_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4479_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4479_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4149_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4149_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4429_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4429_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4412_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4412_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4471_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4471_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4459_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4459_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4123_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4123_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4294_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4294_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4331_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4331_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4303_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4303_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4108_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4108_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4263_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4263_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4481_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4481_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4233_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4233_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4148_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4148_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4310_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4310_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4369_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4369_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4129_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4129_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4411_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4411_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4427_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4427_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4428_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4428_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4267_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4267_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4102_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4102_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4478_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4478_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4376_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4376_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4196_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4196_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4420_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4420_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4309_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4309_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4143_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4143_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4269_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4269_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4464_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4464_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4236_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4236_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4365_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4365_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4106_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4106_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4325_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4325_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4484_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4484_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4417_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4417_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4295_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4295_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4289_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4289_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4407_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4407_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4185_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4185_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4449_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4449_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4431_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4431_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4142_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4142_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4474_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4474_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4103_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4103_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4314_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4314_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4329_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4329_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4356_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4356_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4344_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4344_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4447_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4447_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4144_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4144_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4270_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4270_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4181_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4181_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4139_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4139_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4166_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4166_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4191_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4191_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4320_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4320_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4171_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4171_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4274_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4274_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4340_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4340_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4463_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4463_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4381_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4381_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4351_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4351_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4371_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4371_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4264_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4264_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4346_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4346_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4384_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4384_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4409_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4409_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4328_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4328_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4388_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4388_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4453_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4453_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4224_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4224_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4279_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4279_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4235_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4235_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4197_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4197_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4296_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4296_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4122_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4122_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4182_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4182_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4330_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4330_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4298_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4298_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4140_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4140_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4462_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4462_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4437_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4437_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4311_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4311_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4247_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4247_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4131_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4131_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4415_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4415_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4187_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4187_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4404_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4404_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4401_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4401_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4126_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4126_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4307_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4307_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4251_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4251_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4432_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4432_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4146_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4146_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4475_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4475_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4245_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4245_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4287_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4287_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4109_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4109_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4450_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4450_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4111_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4111_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4458_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4458_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4282_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4282_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4413_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4413_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4380_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4380_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4391_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4391_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4155_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4155_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4256_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4256_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4297_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4297_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4436_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4436_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4199_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4199_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4364_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4364_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4249_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4249_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4305_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4305_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4164_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4164_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4306_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4306_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4259_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4259_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4138_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4138_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4435_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4435_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4398_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4398_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4260_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4260_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4156_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4156_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4237_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4237_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4317_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4317_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4332_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4332_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4186_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4186_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4116_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4116_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4211_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4211_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4176_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4176_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4288_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4288_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4221_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4221_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4177_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4177_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4190_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4190_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4336_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4336_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4337_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4337_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4183_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4183_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4399_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4399_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4229_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4229_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4343_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4343_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4262_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4262_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4125_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4125_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4276_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4276_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4292_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4292_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4280_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4280_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4358_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4358_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4286_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4286_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4461_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4461_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4124_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4124_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4253_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4253_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4145_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4145_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4238_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4238_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4290_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4290_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4230_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4230_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4455_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4455_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4322_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4322_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4104_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4104_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4385_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4385_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4468_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4468_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4304_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4304_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4326_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4326_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4451_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4451_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4242_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4242_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4157_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4157_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4355_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4355_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4141_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4141_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4152_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4152_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4284_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4284_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4452_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4452_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4440_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4440_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4266_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4266_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4472_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4472_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4448_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4448_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4308_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4308_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4405_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4405_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4283_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4283_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4128_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4128_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4442_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4442_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4252_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4252_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4433_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4433_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4419_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4419_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4316_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4316_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4273_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4273_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4319_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4319_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4261_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4261_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4430_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4430_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4172_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4172_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4254_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4254_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4366_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4366_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4150_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4150_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4248_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4248_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4255_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4255_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4315_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4315_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4439_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4439_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4342_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4342_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4225_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4225_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4466_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4466_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4213_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4213_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4313_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4313_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4173_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4173_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4387_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4387_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4302_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4302_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4318_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4318_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4188_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4188_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4250_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4250_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4206_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4206_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4107_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4107_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4291_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4291_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4467_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4467_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4217_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4217_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4359_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4359_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4133_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4133_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4425_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4425_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4293_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4293_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4209_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4209_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4473_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4473_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4400_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4400_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4194_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4194_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4216_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4216_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4208_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4208_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4203_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4203_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4151_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4151_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4285_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4285_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4389_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4389_L001_L002_L003_L004.fa\n",
      "Transferring /home/al/analysis/11-16.11.2020/consensus_sequences/illumina/SEARCH-4300_L001_L002_L003_L004.fa to /home/al/analysis/ivar_retrospect/consensus/SEARCH-4300_L001_L002_L003_L004.fa\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_dir = \"/home/al/analysis/ivar_retrospect/consensus/\"\n",
    "rename_fasta_hdrs(out_dir, new_cns_filepaths, hdr_suffix='_new')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "def concat_fasta(in_dir, out_dir):\n",
    "    \"\"\"Concatenate fasta sequences into single fasta file\"\"\"\n",
    "    cat_cmd = f\"cat {in_dir}/*.fa* > {out_dir}.fa\"\n",
    "    subprocess.check_call(cat_cmd, shell=True)\n",
    "    return f\"{out_dir}.fa\"\n",
    "in_dir = \"/home/al/analysis/ivar_retrospect/consensus/\"\n",
    "out_dir = \"/home/al/analysis/ivar_retrospect/cns\"\n",
    "fasta_fp = concat_fasta(in_dir, out_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/al/analysis/ivar_retrospect/cns.fa'"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta_fp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/al/analysis/ivar_retrospect/cns_aligned.fa'"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def align_fasta(fasta_filepath, num_cpus=8, ref_fp: str=''):\n",
    "    \"\"\"Generate Multiple Sequence Alignment of concatenated sequences in input fasta file using mafft\"\"\"\n",
    "    out_filepath = fasta_filepath.split('.')[0] + '_aligned.fa'\n",
    "    msa_cmd = f\"mafft --auto --thread {num_cpus} --keeplength --addfragments {fasta_filepath} {ref_fp} > {out_filepath}\"\n",
    "    subprocess.check_call(msa_cmd, shell=True)\n",
    "    return out_filepath\n",
    "ref_fp = '/home/al/data/hcov19/NC045512.fasta'\n",
    "msa_fp = align_fasta(fasta_fp, ref_fp=ref_fp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_tree(msa_filepath, num_cpus=8):\n",
    "    \"\"\"Compute ML tree of aligned sequences in input fasta using iqtree\"\"\"\n",
    "    out_filepath = msa_filepath + '.treefile'\n",
    "    tree_cmd = f\"iqtree -s {msa_filepath} -nt 6 -m HKY -czb -fast\"\n",
    "    subprocess.check_call(tree_cmd, shell=True)\n",
    "    return out_filepath"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/al/analysis/ivar_retrospect/cns_aligned_tree'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compute_tree('/home/al/analysis/ivar_retrospect/cns_aligned.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/al/analysis/ivar_retrospect/timetree'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def compute_time_tree(msa_filepath, tree_filepath, num_cpus=8):\n",
    "    out_path = '/'.join(msa_filepath.split('/')[:-1]) + '/timetree'\n",
    "    tree_cmd = f\"treetime ancestral --aln {msa_filepath} --tree {tree_filepath} --outdir {out_path}\"\n",
    "    subprocess.check_call(tree_cmd, shell=True)\n",
    "    return out_path\n",
    "tree_filepath = '/home/al/analysis/ivar_retrospect/cns_aligned.fa.treefile'\n",
    "msa_filepath = '/home/al/analysis/ivar_retrospect/cns_aligned.fa'\n",
    "compute_time_tree(msa_filepath, tree_filepath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('json_input.csv').sample(2)\n",
    "df.to_csv('json_input.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(100, 46)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.sample(100, replace=True).shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequence Release"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = pd.read_csv(\"/home/al/analysis/2020-12-01_release_2/metadata.csv\")\n",
    "# df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#!/bin/bash\n",
      "\n",
      "python3 bjorn.py --not-dry-run --out-dir /home/al/analysis/bjorn_test --sample-sheet /home/al/data/COVID_sequencing_summary-GISAID.csv --cpus 8 --analysis-folder /home/gk/analysis/ --output-metadata /home/al/code/HCoV-19-Genomics/metadata.csv\n"
     ]
    }
   ],
   "source": [
    "!cat run.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# path to reference sequence (used later for MSA and tree construction)\n",
    "ref_path = Path('/home/gk/code/hCoV19/db/NC045512.fasta')\n",
    "# whether or not to include bam files in the release\n",
    "include_bams = False\n",
    "# path to reference sequence (used later for MSA and tree construction)\n",
    "# this is the directory where results get saved\n",
    "out_dir = Path('/home/al/analysis/bjorn_test')\n",
    "# number of cores to use\n",
    "num_cpus = 4\n",
    "# file path to samples sheet (make sure it is the most recent)\n",
    "sample_sheet_fpath = Path('/home/al/data/COVID_sequencing_summary-GISAID.csv')\n",
    "# path to analysis results\n",
    "analysis_fpath = Path('/home/gk/analysis/')\n",
    "# file path to metadata of samples that have already been released\n",
    "released_samples_fpath = Path('/home/al/code/HCoV-19-Genomics/metadata.csv')\n",
    "# Whether run is dry\n",
    "dry_run = True\n",
    "# COLUMNS TO INCLUDE IN METADATA\n",
    "git_meta_cols = [\"ID\", \"collection_date\", \"location\", \n",
    "                 \"percent_coverage_cds\", \"avg_depth\", \n",
    "                 \"authors\", \"originating_lab\"]\n",
    "gisaid_meta_cols = ['Submitter',\n",
    "                   'FASTA filename', 'Virus name', 'Type', 'Passage details/history',\n",
    "                   'Collection date', 'location', 'Additional location information',\n",
    "                   'Host', 'Additional host information', 'Gender', 'Patient age',\n",
    "                   'Patient status', 'Specimen source', 'Outbreak', 'Last vaccinated',\n",
    "                   'Treatment', 'Sequencing technology', 'Assembly method', 'Coverage',\n",
    "                   'originating_lab', 'Address', 'Sample ID given by the sample provider',\n",
    "                   'Submitting lab', 'Address.1',\n",
    "                   'Sample ID given by the submitting laboratory', 'authors', 'avg_depth']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Dry run: True.\n",
      "Include BAMS: False.\n",
      "Reading release summary file from /home/al/data/COVID_sequencing_summary-GISAID.csv.\n",
      "Reading repository metadata from /home/al/code/HCoV-19-Genomics/metadata.csv.\n",
      "Searching analysis folder /home/gk/analysis/.\n",
      "\n",
      "Preparing 366 samples for release\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "# # Test\n",
    "# out_dir = \"/home/gk/southpark/2020-11-21_release\"\n",
    "# sample_sheet_fpath = \"/home/gk/code/hCoV19/release_summary_csv/2020-11-20_seq_summary.csv\"\n",
    "# analysis_fpath = \"/home/gk/analysis/\"\n",
    "# released_samples_fpath = \"/home/gk/analysis/hcov-19-genomics/metadata.csv\"\n",
    "# dry_run = True\n",
    "\n",
    "print(f\"\"\"\n",
    "Dry run: {dry_run}.\n",
    "Include BAMS: {include_bams}.\n",
    "Reading release summary file from {sample_sheet_fpath}.\n",
    "Reading repository metadata from {released_samples_fpath}.\n",
    "Searching analysis folder {analysis_fpath}.\n",
    "\"\"\"\n",
    ")\n",
    "\n",
    "# Collecting Sequence Data\n",
    "\n",
    "\n",
    "# grab all filepaths for bam data\n",
    "bam_filepaths = glob.glob(f\"{analysis_fpath}/**/merged_aligned_bams/illumina/*.bam\")\n",
    "bam_filepaths = [Path(fp) for fp in bam_filepaths]\n",
    "# consolidate sample ID format\n",
    "bam_ids = get_ids(bam_filepaths)\n",
    "# Turn into dataframe\n",
    "bam_data = list(zip(*[bam_ids, bam_filepaths]))\n",
    "bam_df = pd.DataFrame(data=bam_data, columns=['sample_id', 'PATH'])\n",
    "# grab all paths to consensus sequences\n",
    "consensus_filepaths = glob.glob(f\"{analysis_fpath}/**/consensus_sequences/illumina/*.fa\")\n",
    "consensus_filepaths = [Path(fp) for fp in consensus_filepaths]\n",
    "# consolidate sample ID format\n",
    "consensus_ids = get_ids(consensus_filepaths)\n",
    "# Turn into dataframe\n",
    "consensus_data = list(zip(*[consensus_ids, consensus_filepaths]))\n",
    "consensus_df = pd.DataFrame(data=consensus_data, columns=['sample_id', 'PATH'])\n",
    "# clean up cns and bam (remove duplicate IDs)\n",
    "bam_df.drop_duplicates(subset=['sample_id'], keep='last', inplace=True)\n",
    "consensus_df.drop_duplicates(subset=['sample_id'], keep='last', inplace=True)\n",
    "# include only SEARCH samples\n",
    "consensus_df = consensus_df[(consensus_df['sample_id'].str.contains('SEARCH'))]\n",
    "# merge consensus and bam filepaths for each sample ID\n",
    "analysis_df = pd.merge(consensus_df, bam_df, on='sample_id', how='left')\n",
    "# load sample sheet data (GISAID) - make sure to download most recent one\n",
    "seqsum = pd.read_csv(sample_sheet_fpath)\n",
    "# clean up\n",
    "seqsum = seqsum[(~seqsum['SEARCH SampleID'].isna()) & (seqsum['SEARCH SampleID']!='#REF!')]\n",
    "# consolidate sample ID format\n",
    "seqsum.loc[:, 'sample_id'] = seqsum['SEARCH SampleID'].apply(process_id)\n",
    "seqsum.drop_duplicates(subset=['sample_id'], keep='last', inplace=True)\n",
    "seqsum = seqsum[seqsum['New sequences ready for release'] == 'Yes']\n",
    "num_seqs_to_release = seqsum['sample_id'].unique().shape[0]\n",
    "# JOIN summary sheet with analysis meta data\n",
    "sequence_results = pd.merge(seqsum, analysis_df, on='sample_id', how='inner')\n",
    "# compute number of samples with missing consensus and/or bam files\n",
    "num_seqs_found = sequence_results['sample_id'].unique().shape[0]\n",
    "num_samples_missing_cons = num_seqs_to_release - num_seqs_found\n",
    "num_samples_missing_bams = 'NA'\n",
    "if include_bams:\n",
    "    # exclude any samples that do not have BAM data\n",
    "    num_samples_missing_bams = sequence_results[sequence_results['PATH_y'].isna()].shape[0]\n",
    "    sequence_results = sequence_results[~sequence_results['PATH_y'].isna()]\n",
    "# samples missing consensus or BAM sequence files\n",
    "# ## Make sure to remove any samples that have already been uploaded to github (just an extra safety step)\n",
    "# load metadata.csv from github repo, then clean up\n",
    "meta_df = pd.read_csv(released_samples_fpath)\n",
    "meta_df = meta_df[meta_df['ID'].str.contains('SEARCH')]\n",
    "# consolidate sample ID format\n",
    "meta_df.loc[:, 'sample_id'] = meta_df['ID'].apply(process_id)\n",
    "# get IDs of samples that have already been released\n",
    "released_seqs = meta_df['sample_id'].unique()\n",
    "# filter out released samples from all the samples we got\n",
    "final_result = sequence_results.copy()\n",
    "# final_result = sequence_results[~sequence_results['sample_id'].isin(released_seqs)]\n",
    "print(f\"Preparing {final_result.shape[0]} samples for release\")\n",
    "# Transfer files\n",
    "if not dry_run:\n",
    "    transfer_files(final_result, out_dir, include_bams=include_bams, ncpus=num_cpus)\n",
    "# ## Getting coverage information\n",
    "cov_filepaths = glob.glob(f\"{analysis_fpath}/**/trimmed_bams/illumina/reports/*.tsv\")\n",
    "cov_filepaths = [Path(fp) for fp in cov_filepaths]\n",
    "# read coverage data and clean it up\n",
    "cov_df = pd.concat((pd.read_csv(f, sep='\\t').assign(path=f) for f in cov_filepaths))\n",
    "cov_df.loc[:,'sample_id'] = cov_df['SAMPLE'].apply(process_coverage_sample_ids)\n",
    "cov_df.loc[:,'date'] = cov_df['path'].apply(lambda x: ''.join(x.split('/')[4].split('.')[:3]))\n",
    "cov_df = (cov_df.sort_values('date')\n",
    "                .drop_duplicates(subset=['sample_id'], keep='last'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ans.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Transfer Complete. All results saved in /home/al/analysis/bjorn_test\n"
     ]
    }
   ],
   "source": [
    "# JOIN results with coverage info\n",
    "ans = (\n",
    "    pd.merge(final_result, cov_df, on='sample_id', how='left')\n",
    "    .assign(\n",
    "        collection_date = lambda x: pd.to_datetime(x[\"Collection date\"]).dt.strftime(\"%Y-%m-%d\")\n",
    "    )\n",
    "    .rename(columns={\n",
    "        \"SEARCH SampleID\": \"ID\",\n",
    "        \"Location\": \"location\",\n",
    "        \"COVERAGE\": \"percent_coverage_cds\",\n",
    "        \"AVG_DEPTH\": \"avg_depth\",\n",
    "        \"Authors\": \"authors\",\n",
    "        \"Originating lab\": \"originating_lab\"\n",
    "    })\n",
    ")\n",
    "num_samples_missing_coverage = ans[ans['percent_coverage_cds'].isna()].shape[0]\n",
    "\n",
    "def create_github_meta(new_meta_df: pd.DataFrame, old_meta_filepath: str, meta_cols: list):\n",
    "    \"\"\"Generate Github metadata with updated information about newly released samples\"\"\"\n",
    "    old_metadata = pd.read_csv(old_meta_filepath)\n",
    "    new_metadata = pd.concat([old_metadata, new_meta_df.loc[:, meta_cols]])\n",
    "    new_metadata.to_csv(out_dir/'metadata.csv', index=False)\n",
    "    return f\"Github metadata saved in {out_dir/'metadata.csv'}\"\n",
    "\n",
    "def create_gisaid_meta(new_meta_df: pd.DataFrame, meta_cols: list):\n",
    "    \"\"\"Generate GISAID metadata for newly released samples\"\"\"\n",
    "    new_meta_df[meta_cols].to_csv(out_dir/'gisaid_metadata.csv', index=False)\n",
    "    return f\"GISAID metadata saved in {out_dir/'gisaid_metadata.csv'}\"\n",
    "\n",
    "create_github_meta(ans, released_samples_fpath, git_meta_cols)\n",
    "create_gisaid_meta(ans, gisaid_meta_cols) \n",
    "low_coverage_samples = ans[ans[\"percent_coverage_cds\"] < 90]\n",
    "# Data logging\n",
    "with open(\"{}/data_release.log\".format(out_dir), 'w') as f:\n",
    "    f.write(f'{num_samples_missing_coverage} samples are missing coverage information\\n')\n",
    "    f.write(f'{low_coverage_samples.shape[0]} samples were found to have coverage below 90%\\n')\n",
    "    f.write(f'{num_samples_missing_cons} samples were ignored because they were missing consensus sequence files\\n')\n",
    "    f.write(f'{num_samples_missing_bams} samples were ignored because they were missing BAM sequence files\\n')\n",
    "print(f\"Transfer Complete. All results saved in {out_dir}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "243"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "num_samples_missing_coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Sample ID', 'ID', 'Ready for release?',\n",
       "       'New sequences ready for release', 'Released?', 'Submitter',\n",
       "       'FASTA filename', 'Virus name', 'Type', 'Passage details/history',\n",
       "       'Collection date', 'location', 'Additional location information',\n",
       "       'Host', 'Additional host information', 'Gender', 'Patient age',\n",
       "       'Patient status', 'Specimen source', 'Outbreak', 'Last vaccinated',\n",
       "       'Treatment', 'Sequencing technology', 'Assembly method', 'Coverage',\n",
       "       'originating_lab', 'Address', 'Sample ID given by the sample provider',\n",
       "       'Submitting lab', 'Address.1',\n",
       "       'Sample ID given by the submitting laboratory', 'authors', 'Comment',\n",
       "       'Comment Icon', 'sample_id', 'PATH_x', 'PATH_y', 'SAMPLE',\n",
       "       'percent_coverage_cds', 'avg_depth', 'MIN', 'MAX', 'ZERO_DEPTH', 'path',\n",
       "       'date', 'collection_date'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ans.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Genbank"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "cons_seqs = Path(out_dir/'fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "msa_dir = out_dir/'msa'\n",
    "if not Path.isdir(msa_dir):\n",
    "    Path.mkdir(msa_dir);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Path('/home/al/analysis/bjorn_test/fa/NC045512.fasta')"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cons_seqs = Path(out_dir/'fa')\n",
    "copy(ref_path, cons_seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Path('bjorn_test')"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out_dir.basename()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# concat_fasta(out_dir, msa_dir/out_dir.basename())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bjorn_test_aligned.fa  bjorn_test.fa\n"
     ]
    }
   ],
   "source": [
    "!ls {msa_dir}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cns_seqs = SeqIO.parse(msa_dir/out_dir.basename()+'.fa', 'fasta')\n",
    "# cns_seqs = list(cns_seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_genbank_meta(df: pd.DataFrame, meta_cols: list) -> pd.DataFrame:\n",
    "    genbank_meta = df[meta_cols].copy()\n",
    "    genbank_meta['country'] = genbank_meta['location'].apply(lambda x: x.split('/')[0])\n",
    "    genbank_meta['isolate'] = genbank_meta['Virus name'].str.replace('hCoV-19', 'SARS-CoV-2/human')\n",
    "    genbank_meta['host'] = genbank_meta['Host'].str.replace('Human', 'Homo Sapiens')\n",
    "    genbank_meta.rename(columns={'Virus name': 'Sequence_ID', 'collection_date': 'collection-date',\n",
    "                                 'Specimen source': 'isolation-source'}, inplace=True)\n",
    "    genbank_meta.loc[genbank_meta['country']=='MEX', 'country'] = 'Mexico'\n",
    "    return genbank_meta[['Sequence_ID', 'isolate', 'country', \n",
    "                         'collection-date', 'host', 'isolation-source']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def assemble_genbank_release(cns_seqs: list, df: pd.DataFrame, meta_cols: list):\n",
    "    # create directory for genbank release\n",
    "    genbank_dir = out_dir/'genbank'\n",
    "    if not Path.isdir(genbank_dir):\n",
    "        Path.mkdir(genbank_dir);\n",
    "    authors = {}\n",
    "    # group samples by author\n",
    "    for ctr, (n, grp) in enumerate(df.groupby('authors')):\n",
    "        authors[ctr+1] = n\n",
    "        # generate sample metadata\n",
    "        genbank_meta = create_genbank_meta(grp, meta_cols)\n",
    "        genbank_meta.to_csv(genbank_dir/f'genbank_metadata_{ctr+1}.tsv', sep='\\t', index=False)\n",
    "        # fetch consensus sequences of those samples\n",
    "        recs = [i for i in cns_seqs if i.name in genbank_meta['Sequence_ID'].tolist()]\n",
    "        SeqIO.write(recs, genbank_dir/f'genbank_release_{ctr+1}.fa', 'fasta')\n",
    "    # write mapping of index to author for later reference\n",
    "    (pd.DataFrame.from_dict(authors, orient='index')\n",
    "       .rename(columns={0: 'authors'})\n",
    "       .to_csv(genbank_dir/'authors.tsv', sep='\\t'))\n",
    "    return f\"Genbank data release saved in {genbank_dir}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# genbank_meta_cols = ['Sample ID', 'ID', 'Virus name', 'location', \n",
    "#                         'Specimen source', 'collection_date', 'Host']\n",
    "# assemble_genbank_release(cns_seqs, ans, genbank_meta_cols)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SRA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sra_dir = out_dir/'sra'\n",
    "# if not Path.isdir(sra_dir):\n",
    "#     Path.mkdir(sra_dir);\n",
    "# input(f\"Have you received the BioSample.txt files and placed them inside {sra_dir}? \\n \\\n",
    "# Press Enter to continue...\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_sra_meta(df: pd.DataFrame, sra_dir: Path):\n",
    "    biosample_paths = glob.glob(f\"{sra_dir}/*.txt\")\n",
    "    biosample_df = pd.concat((pd.read_csv(f, sep='\\t') for f in biosample_paths))\n",
    "    biosample_df[\"sample_id\"] = biosample_df[\"Sample Name\"].apply(lambda x: \"\".join(x.split(\"-\")[:2]))\n",
    "    bam_files = ans.loc[~ans['PATH_y'].isna()][['sample_id', 'PATH_y']].rename(columns={'PATH_y': 'file_name'})\n",
    "    sra_merged = pd.merge(biosample_df, bam_files, on=\"sample_id\")\n",
    "    sra_merged[[\"Accession\", \"Sample Name\", \"file_name\"]].to_csv(sra_dir/\"sra_metadata.csv\", index=False)\n",
    "    return f\"SRA metadata saved in {sra_dir/'sra_metadata.csv'}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TEST behavior\n",
    "# cat_cmd = f\"cat {out_dir}/fa/*.fa > {msa_dir/out_dir.basename()}.fa\"\n",
    "# subprocess.check_call(cat_cmd, shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## INDELS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# msa_fp = align_fasta(msa_dir/out_dir.basename())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "msa_fp = msa_dir/'bjorn_test_aligned.fa'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "from onion_trees import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "patient_zero = 'NC_045512.2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Bio.Align.MultipleSeqAlignment"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read MSA file\n",
    "cns_seqs = AlignIO.read(msa_fp, 'fasta')\n",
    "type(cns_seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqs = get_seqs(cns_seqs)\n",
    "seqsdf = (pd.DataFrame(index=seqs.keys(), data=seqs.values(), columns=['sequence'])\n",
    "            .reset_index().rename(columns={'index': 'idx'}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'atccgaacaggccacaaatgagtt'"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqsdf.loc[seqsdf['idx'].str.contains('4650'), 'sequence'].values[0][25890:25914]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'atccg--------------gagtt'"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqsdf.loc[seqsdf['idx'].str.contains('NC'), 'sequence'].values[0][25890:25914]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relative_coords</th>\n",
       "      <th>ins_len</th>\n",
       "      <th>samples</th>\n",
       "      <th>num_samples</th>\n",
       "      <th>type</th>\n",
       "      <th>absolute_coords</th>\n",
       "      <th>prev_5nts</th>\n",
       "      <th>next_5nts</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>25895:25908</td>\n",
       "      <td>14</td>\n",
       "      <td>[hCoV-19/USA/SEARCH-4650-SAN/2020]</td>\n",
       "      <td>1</td>\n",
       "      <td>insertion</td>\n",
       "      <td>26160:26173</td>\n",
       "      <td>atccg</td>\n",
       "      <td>gagtt</td>\n",
       "      <td>ORF3a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10811:10813</td>\n",
       "      <td>3</td>\n",
       "      <td>[hCoV-19/USA/SEARCH-4288-SAN/2020, hCoV-19/USA...</td>\n",
       "      <td>2</td>\n",
       "      <td>insertion</td>\n",
       "      <td>11076:11078</td>\n",
       "      <td>tgttc</td>\n",
       "      <td>ttttt</td>\n",
       "      <td>ORF1ab</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  relative_coords  ins_len                                            samples  \\\n",
       "1     25895:25908       14                 [hCoV-19/USA/SEARCH-4650-SAN/2020]   \n",
       "0     10811:10813        3  [hCoV-19/USA/SEARCH-4288-SAN/2020, hCoV-19/USA...   \n",
       "\n",
       "   num_samples       type absolute_coords prev_5nts next_5nts    gene  \n",
       "1            1  insertion     26160:26173     atccg     gagtt   ORF3a  \n",
       "0            2  insertion     11076:11078     tgttc     ttttt  ORF1ab  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inserts = identify_insertions(msa_fp, patient_zero)\n",
    "inserts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'tctta---acaac'"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqsdf.loc[seqsdf['idx'].str.contains('4351'), 'sequence'].values[0][1334:1347]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relative_coords</th>\n",
       "      <th>del_len</th>\n",
       "      <th>samples</th>\n",
       "      <th>num_samples</th>\n",
       "      <th>type</th>\n",
       "      <th>absolute_coords</th>\n",
       "      <th>prev_5nts</th>\n",
       "      <th>next_5nts</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1339:1341</td>\n",
       "      <td>3</td>\n",
       "      <td>[hCoV-19/USA/SEARCH-4351-SAN/2020]</td>\n",
       "      <td>1</td>\n",
       "      <td>deletion</td>\n",
       "      <td>1604:1606</td>\n",
       "      <td>tctta</td>\n",
       "      <td>acaac</td>\n",
       "      <td>ORF1ab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>21716:21727</td>\n",
       "      <td>12</td>\n",
       "      <td>[hCoV-19/USA/SEARCH-4199-SAN/2020]</td>\n",
       "      <td>1</td>\n",
       "      <td>deletion</td>\n",
       "      <td>21981:21992</td>\n",
       "      <td>ccatt</td>\n",
       "      <td>ttacc</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>249:254</td>\n",
       "      <td>6</td>\n",
       "      <td>[hCoV-19/USA/SEARCH-4206-SAN/2020]</td>\n",
       "      <td>1</td>\n",
       "      <td>deletion</td>\n",
       "      <td>514:519</td>\n",
       "      <td>gtcat</td>\n",
       "      <td>gttga</td>\n",
       "      <td>ORF1ab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>251:254</td>\n",
       "      <td>4</td>\n",
       "      <td>[hCoV-19/USA/SEARCH-4449-SAN/2020]</td>\n",
       "      <td>1</td>\n",
       "      <td>deletion</td>\n",
       "      <td>516:519</td>\n",
       "      <td>catgt</td>\n",
       "      <td>gttga</td>\n",
       "      <td>ORF1ab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>25894:25901</td>\n",
       "      <td>8</td>\n",
       "      <td>[hCoV-19/USA/SEARCH-4264-SAN/2020]</td>\n",
       "      <td>1</td>\n",
       "      <td>deletion</td>\n",
       "      <td>26159:26166</td>\n",
       "      <td>gttgt</td>\n",
       "      <td>taatg</td>\n",
       "      <td>ORF3a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>27818:27844</td>\n",
       "      <td>27</td>\n",
       "      <td>[hCoV-19/USA/SEARCH-4298-SAN/2020]</td>\n",
       "      <td>1</td>\n",
       "      <td>deletion</td>\n",
       "      <td>28083:28109</td>\n",
       "      <td>ggatg</td>\n",
       "      <td>acatc</td>\n",
       "      <td>ORF8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>27824:27829</td>\n",
       "      <td>6</td>\n",
       "      <td>[hCoV-19/USA/SEARCH-4583-SAN/2020]</td>\n",
       "      <td>1</td>\n",
       "      <td>deletion</td>\n",
       "      <td>28089:28094</td>\n",
       "      <td>ggctg</td>\n",
       "      <td>aatca</td>\n",
       "      <td>ORF8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>28615:28632</td>\n",
       "      <td>18</td>\n",
       "      <td>[hCoV-19/USA/SEARCH-4145-SAN/2020]</td>\n",
       "      <td>1</td>\n",
       "      <td>deletion</td>\n",
       "      <td>28880:28897</td>\n",
       "      <td>cagta</td>\n",
       "      <td>gaatg</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>25892:25895</td>\n",
       "      <td>4</td>\n",
       "      <td>[hCoV-19/USA/SEARCH-4355-SAN/2020, hCoV-19/USA...</td>\n",
       "      <td>2</td>\n",
       "      <td>deletion</td>\n",
       "      <td>26157:26160</td>\n",
       "      <td>gagtt</td>\n",
       "      <td>atcca</td>\n",
       "      <td>ORF3a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>27421:27444</td>\n",
       "      <td>24</td>\n",
       "      <td>[hCoV-19/USA/SEARCH-4225-SAN/2020, hCoV-19/USA...</td>\n",
       "      <td>2</td>\n",
       "      <td>deletion</td>\n",
       "      <td>27686:27709</td>\n",
       "      <td>tactc</td>\n",
       "      <td>aatag</td>\n",
       "      <td>ORF7a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>27484:27491</td>\n",
       "      <td>8</td>\n",
       "      <td>[hCoV-19/USA/SEARCH-4549-SAN/2020, hCoV-19/USA...</td>\n",
       "      <td>2</td>\n",
       "      <td>deletion</td>\n",
       "      <td>27749:27756</td>\n",
       "      <td>agaaa</td>\n",
       "      <td>gattg</td>\n",
       "      <td>ORF7a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21715:21723</td>\n",
       "      <td>9</td>\n",
       "      <td>[hCoV-19/USA/SEARCH-4531-SAN/2020, hCoV-19/USA...</td>\n",
       "      <td>3</td>\n",
       "      <td>deletion</td>\n",
       "      <td>21980:21988</td>\n",
       "      <td>tccat</td>\n",
       "      <td>tttat</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>28624:28635</td>\n",
       "      <td>12</td>\n",
       "      <td>[hCoV-19/USA/SEARCH-4138-SAN/2020, hCoV-19/USA...</td>\n",
       "      <td>3</td>\n",
       "      <td>deletion</td>\n",
       "      <td>28889:28900</td>\n",
       "      <td>aactt</td>\n",
       "      <td>tggct</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>27613:27625</td>\n",
       "      <td>13</td>\n",
       "      <td>[hCoV-19/USA/SEARCH-4415-SAN/2020, hCoV-19/USA...</td>\n",
       "      <td>6</td>\n",
       "      <td>deletion</td>\n",
       "      <td>27878:27890</td>\n",
       "      <td>cttgt</td>\n",
       "      <td>acatg</td>\n",
       "      <td>ORF7b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>250:252</td>\n",
       "      <td>3</td>\n",
       "      <td>[hCoV-19/USA/SEARCH-4342-SAN/2020, hCoV-19/USA...</td>\n",
       "      <td>10</td>\n",
       "      <td>deletion</td>\n",
       "      <td>515:517</td>\n",
       "      <td>tcatg</td>\n",
       "      <td>tggtt</td>\n",
       "      <td>ORF1ab</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   relative_coords  del_len  \\\n",
       "0        1339:1341        3   \n",
       "2      21716:21727       12   \n",
       "3          249:254        6   \n",
       "5          251:254        4   \n",
       "7      25894:25901        8   \n",
       "11     27818:27844       27   \n",
       "12     27824:27829        6   \n",
       "13     28615:28632       18   \n",
       "6      25892:25895        4   \n",
       "8      27421:27444       24   \n",
       "9      27484:27491        8   \n",
       "1      21715:21723        9   \n",
       "14     28624:28635       12   \n",
       "10     27613:27625       13   \n",
       "4          250:252        3   \n",
       "\n",
       "                                              samples  num_samples      type  \\\n",
       "0                  [hCoV-19/USA/SEARCH-4351-SAN/2020]            1  deletion   \n",
       "2                  [hCoV-19/USA/SEARCH-4199-SAN/2020]            1  deletion   \n",
       "3                  [hCoV-19/USA/SEARCH-4206-SAN/2020]            1  deletion   \n",
       "5                  [hCoV-19/USA/SEARCH-4449-SAN/2020]            1  deletion   \n",
       "7                  [hCoV-19/USA/SEARCH-4264-SAN/2020]            1  deletion   \n",
       "11                 [hCoV-19/USA/SEARCH-4298-SAN/2020]            1  deletion   \n",
       "12                 [hCoV-19/USA/SEARCH-4583-SAN/2020]            1  deletion   \n",
       "13                 [hCoV-19/USA/SEARCH-4145-SAN/2020]            1  deletion   \n",
       "6   [hCoV-19/USA/SEARCH-4355-SAN/2020, hCoV-19/USA...            2  deletion   \n",
       "8   [hCoV-19/USA/SEARCH-4225-SAN/2020, hCoV-19/USA...            2  deletion   \n",
       "9   [hCoV-19/USA/SEARCH-4549-SAN/2020, hCoV-19/USA...            2  deletion   \n",
       "1   [hCoV-19/USA/SEARCH-4531-SAN/2020, hCoV-19/USA...            3  deletion   \n",
       "14  [hCoV-19/USA/SEARCH-4138-SAN/2020, hCoV-19/USA...            3  deletion   \n",
       "10  [hCoV-19/USA/SEARCH-4415-SAN/2020, hCoV-19/USA...            6  deletion   \n",
       "4   [hCoV-19/USA/SEARCH-4342-SAN/2020, hCoV-19/USA...           10  deletion   \n",
       "\n",
       "   absolute_coords prev_5nts next_5nts    gene  \n",
       "0        1604:1606     tctta     acaac  ORF1ab  \n",
       "2      21981:21992     ccatt     ttacc       S  \n",
       "3          514:519     gtcat     gttga  ORF1ab  \n",
       "5          516:519     catgt     gttga  ORF1ab  \n",
       "7      26159:26166     gttgt     taatg   ORF3a  \n",
       "11     28083:28109     ggatg     acatc    ORF8  \n",
       "12     28089:28094     ggctg     aatca    ORF8  \n",
       "13     28880:28897     cagta     gaatg       N  \n",
       "6      26157:26160     gagtt     atcca   ORF3a  \n",
       "8      27686:27709     tactc     aatag   ORF7a  \n",
       "9      27749:27756     agaaa     gattg   ORF7a  \n",
       "1      21980:21988     tccat     tttat       S  \n",
       "14     28889:28900     aactt     tggct       N  \n",
       "10     27878:27890     cttgt     acatg   ORF7b  \n",
       "4          515:517     tcatg     tggtt  ORF1ab  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dels = identify_deletions(msa_fp, patient_zero)\n",
    "dels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_seq = get_seq(cns_seqs, patient_zero)[265:29674]\n",
    "insert_positions = identify_insertion_positions(ref_seq)\n",
    "seqs = get_seqs(cns_seqs)\n",
    "seqsdf = (pd.DataFrame(index=seqs.keys(), data=seqs.values(), columns=['sequence'])\n",
    "            .reset_index().rename(columns={'index': 'idx'}))\n",
    "# seqsdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# seqsdf.iloc[2]['sequence'][10811:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# insert_positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# min_ins_len = 1\n",
    "# def find_insertions(x):\n",
    "#     ins_positions = [m for m in insert_positions if x[m]!='-' and x[m]!='n']\n",
    "#     insertions = [list(insert) for insert in mit.consecutive_groups(ins_positions)]\n",
    "#     return insertions\n",
    "# seqsdf['ins_positions'] = seqsdf['sequence'].apply(find_insertions)\n",
    "# ins_seqs = seqsdf.loc[seqsdf['ins_positions'].str.len() > 0]\n",
    "# ins_seqs = ins_seqs.explode('ins_positions')\n",
    "# ins_seqs['ins_len'] = ins_seqs['ins_positions'].apply(len)\n",
    "# # only consider deletions longer than 2nts\n",
    "# ins_seqs = ins_seqs[ins_seqs['ins_len'] > min_ins_len]\n",
    "# # fetch coordinates of each deletion\n",
    "# ins_seqs['ins_coords'] = ins_seqs['ins_positions'].apply(get_coords)\n",
    "# # group sample by the deletion they share\n",
    "# ins_seqs = (ins_seqs.groupby(['ins_coords', 'ins_len'])\n",
    "#                     .agg(samples=('idx', 'unique'),       # list of sample IDs with the deletion\n",
    "#                          num_samples=('idx', 'nunique'))  # num of samples with the deletion\n",
    "#                     .reset_index()\n",
    "#                     .sort_values('num_samples'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ins_coords</th>\n",
       "      <th>ins_len</th>\n",
       "      <th>samples</th>\n",
       "      <th>num_samples</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>25895:25908</td>\n",
       "      <td>14</td>\n",
       "      <td>[hCoV-19/USA/SEARCH-4650-SAN/2020]</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10811:10813</td>\n",
       "      <td>3</td>\n",
       "      <td>[hCoV-19/USA/SEARCH-4288-SAN/2020, hCoV-19/USA...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ins_coords  ins_len                                            samples  \\\n",
       "1  25895:25908       14                 [hCoV-19/USA/SEARCH-4650-SAN/2020]   \n",
       "0  10811:10813        3  [hCoV-19/USA/SEARCH-4288-SAN/2020, hCoV-19/USA...   \n",
       "\n",
       "   num_samples  \n",
       "1            1  \n",
       "0            2  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ins_seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_seq = get_seq(cns_seqs, patient_zero)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "for rec in cns_seqs:\n",
    "    if rec.id=='hCoV-19/USA/SEARCH-4206-SAN/2020':\n",
    "        seq = rec.seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'gtcat------gttgagctgg'"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq[509:530]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqs, ref_seq = process_cns_seqs(cns_seqs, patient_zero)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "# deletions = identify_deletions(msa_fp, patient_zero, min_del_len=1)\n",
    "# deletions['del_coords'] = deletions['del_coords'].apply(adjust_coords)\n",
    "# deletions['prev_5nts'] = deletions['del_coords'].apply(lambda x: ref_seq[int(x.split(':')[0])-5:int(x.split(':')[0])])\n",
    "# deletions['next_5nts'] = deletions['del_coords'].apply(lambda x: ref_seq[int(x.split(':')[1])+1:int(x.split(':')[1])+6])\n",
    "# deletions['gene'] = deletions['del_coords'].apply(lambda x: int(x.split(':')[0])).apply(map_gene_to_pos)\n",
    "# deletions.to_csv(out_dir/'deletions.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pd.read_csv(out_dir/'deletions.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bjorn_test_aligned.fa  bjorn_test.fa\n"
     ]
    }
   ],
   "source": [
    "!ls {msa_dir}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TEST behavior\n",
    "# msa_cmd = f\"mafft --auto --thread {num_cpus} {msa_dir/out_dir.basename()}.fa > {msa_dir/out_dir.basename()}_aligned.fa\"\n",
    "# subprocess.check_call(msa_cmd, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ans.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ans.to_csv('json_input.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## JSON metadata for CViSB portal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'includedInDataset': 'sarscov2-virus-seq',\n",
       " 'identifier': 'SEARCH-0289-NBG.fa',\n",
       " 'name': 'SEARCH-0289-NBG.fa',\n",
       " 'contentUrl': 'https://console.cloud.google.com/storage/browser/_details/andersen-lab_hcov-19-genomics/consensus_sequences/SEARCH-0289-NBG.fa',\n",
       " 'additionalType': 'raw data',\n",
       " 'variableMeasured': ['SARS-CoV-2 virus sequence'],\n",
       " 'measurementTechnique': ['Nucleic Acid Sequencing'],\n",
       " 'dateModified': '2020-11-16',\n",
       " 'experimentIDs': ['SEARCH-0289-NBG-sarscov2']}"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# datadownload.json\n",
    "{\"includedInDataset\": \"sarscov2-virus-seq\", \"identifier\": \"SEARCH-0289-NBG.fa\", \n",
    " \"name\": \"SEARCH-0289-NBG.fa\", \n",
    " \"contentUrl\": \"https://console.cloud.google.com/storage/browser/_details/andersen-lab_hcov-19-genomics/consensus_sequences/SEARCH-0289-NBG.fa\", \n",
    " \"additionalType\": \"raw data\", \"variableMeasured\": [\"SARS-CoV-2 virus sequence\"], \n",
    " \"measurementTechnique\": [\"Nucleic Acid Sequencing\"], \n",
    " \"dateModified\": \"2020-11-16\", \"experimentIDs\": [\"SEARCH-0289-NBG-sarscov2\"]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "({'patientID': 'SEARCH-0289-NBG',\n",
       "  'dateModified': '2020-11-16',\n",
       "  'cohort': 'COVID-19',\n",
       "  'outcome': 'unknown',\n",
       "  'alternateIdentifier': ['SEARCH-0289-NBG'],\n",
       "  'infectionYear': 2020,\n",
       "  'publisher': {'@type': 'Organization',\n",
       "   'identifier': 'CViSB',\n",
       "   'url': 'https://cvisb.org/',\n",
       "   'name': 'Center for Viral Systems Biology',\n",
       "   'contactPoint': {'@type': 'ContactPoint',\n",
       "    'contactType': 'technical support',\n",
       "    'url': 'https://cvisb.org/',\n",
       "    'email': 'info@cvisb.org'}},\n",
       "  'citation': [{'name': 'Allison Smither, Gilberto Sabino-Santos, Patricia Snarski, Lilia Melnik, Antoinette Bell, Kaylynn Genemaras, Arnaud Drouin, Dahlene Fusco, Robert Garry with SEARCH Alliance San Diego',\n",
       "    '@type': 'Organization'}],\n",
       "  'updatedBy': 'Julia Mullen',\n",
       "  'country': {'@type': 'Country',\n",
       "   'name': 'United States',\n",
       "   'administrativeUnit': 0,\n",
       "   'administrativeType': 'country',\n",
       "   'identifier': 'US',\n",
       "   'url': 'https://www.iso.org/obp/ui/#iso:code:3166:US'},\n",
       "  'presentationDate': '2020-03-31',\n",
       "  'presentationWeek': {'gt': '2020-03-30', 'lt': '2020-04-06'}},)"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# patient.json\n",
    "{\"patientID\": \"SEARCH-0289-NBG\", \n",
    " \"dateModified\": \"2020-11-16\", \n",
    " \"cohort\": \"COVID-19\", \n",
    " \"outcome\": \"unknown\", \n",
    " \"alternateIdentifier\": [\"SEARCH-0289-NBG\"], \n",
    " \"infectionYear\": 2020, \n",
    " \"publisher\": {\"@type\": \"Organization\", \n",
    "               \"identifier\": \"CViSB\", \n",
    "               \"url\": \"https://cvisb.org/\", \n",
    "               \"name\": \"Center for Viral Systems Biology\", \n",
    "               \"contactPoint\": {\"@type\": \"ContactPoint\", \"contactType\": \"technical support\", \n",
    "                                \"url\": \"https://cvisb.org/\", \"email\": \"info@cvisb.org\"}}, \n",
    " \"citation\": [{\"name\": \"Allison Smither, Gilberto Sabino-Santos, Patricia Snarski, Lilia Melnik, Antoinette Bell, Kaylynn Genemaras, Arnaud Drouin, Dahlene Fusco, Robert Garry with SEARCH Alliance San Diego\", \"@type\": \"Organization\"}],\n",
    " \"updatedBy\": \"Julia Mullen\", \n",
    " \"country\": {\"@type\": \"Country\", \"name\": \"United States\", \"administrativeUnit\": 0, \"administrativeType\": \"country\", \"identifier\": \"US\", \"url\": \"https://www.iso.org/obp/ui/#iso:code:3166:US\"}, \"presentationDate\": \"2020-03-31\", \n",
    " \"presentationWeek\": {\"gt\": \"2020-03-30\", \"lt\": \"2020-04-06\"}},"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # experiment.json\n",
    "# {\"experimentID\": \"MG0987-sarscov2\", \n",
    "#  \"privatePatientID\": \"MG0987\", \n",
    "#  \"variableMeasured\": \"SARS-CoV-2 virus sequence\", \n",
    "#  \"measurementTechnique\": \"Nucleic Acid Sequencing\", \n",
    "#  \"includedInDataset\": \"sarscov2-virus-seq\", \n",
    "#  \"dateModified\": \"2020-11-16\", \n",
    "#  \"citation\": [{\"name\": \"SEARCH Alliance San Diego\", \n",
    "#                \"@type\": \"Organization\"}], \n",
    "#  \"data\": [{\"@type\": \"VirusSeqData\", \"virus\": \"SARS-CoV-2\", \n",
    "#            \"virusSegment\": null, \"DNAsequence\": \"NNN\"}]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "def to_json_entry(x):\n",
    "    \"\"\"Return experiment metadata in JSON format compatible with CViSB Data Portal\"\"\"\n",
    "    patient_id = x['ID']\n",
    "    date = x['collection_date']\n",
    "    authors = x['authors']\n",
    "    return {\"experimentID\": f\"{patient_id}-sarscov2\", \"privatePatientID\": patient_id, \n",
    "           \"variableMeasured\": \"SARS-CoV-2 virus sequence\", \n",
    "           \"measurementTechnique\": \"Nucleic Acid Sequencing\", \n",
    "           \"includedInDataset\": \"sarscov2-virus-seq\", \n",
    "           \"dateModified\": date, \n",
    "           \"citation\": [{\"name\": authors, \n",
    "                         \"@type\": \"Organization\"}], \"data\": []}\n",
    "\n",
    "# ans.apply(to_json_entry, axis=1).tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Trees"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## iSNV Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "cols = ['sample', 'GFF_FEATURE', 'POS', 'REF', 'ALT', 'ALT_FREQ']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_ids = [4531, 4690, 4691]\n",
    "v_fps = get_variant_filepaths(sample_ids)\n",
    "v_df = get_variant_data(v_fps)[cols]\n",
    "v_df.drop_duplicates(subset=['sample', 'POS', 'ALT'], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{4531: '/home/gk/analysis/2020.11.25.hCoV19/variants/illumina/SEARCH-4531_L001_L002_L003_L004.tsv',\n",
       " 4690: '/home/gk/analysis/2020.11.25.hCoV19/variants/illumina/SEARCH-4690_L001_L002_L003_L004.tsv',\n",
       " 4691: '/home/gk/analysis/2020.11.25.hCoV19/variants/illumina/SEARCH-4691_L001_L002_L003_L004.tsv'}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v_fps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>GFF_FEATURE</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>ALT_FREQ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>4531</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>12369</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>4690</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>12369</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.976622</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample         GFF_FEATURE    POS REF ALT  ALT_FREQ\n",
       "9     4531  cds-YP_009724389.1  12369   C   T  1.000000\n",
       "27    4690  cds-YP_009724389.1  12369   C   T  0.976622"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v_df[v_df['POS']==12369]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>ALT_FREQ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>4531</td>\n",
       "      <td>12369</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>4690</td>\n",
       "      <td>12369</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.976622</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample    POS REF ALT  ALT_FREQ\n",
       "9     4531  12369   C   T  1.000000\n",
       "27    4690  12369   C   T  0.976622"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v_df.loc[(v_df['REF']=='C')&(v_df['ALT']=='T')&(v_df['POS'] > 12300)&(v_df['POS'] < 12400)][['sample', 'POS', 'REF', 'ALT', 'ALT_FREQ']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "ans = v_df.loc[(v_df.duplicated(subset=['POS'], keep=False))][['sample', 'POS', 'REF', 'ALT', 'ALT_FREQ']].sort_values('POS')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_minor_vars</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4531</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4690</th>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4691</th>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        num_minor_vars\n",
       "sample                \n",
       "4531                 6\n",
       "4690                33\n",
       "4691                14"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v_df['var'] = v_df['POS'].astype(str) + ':' + v_df['ALT']\n",
    "v_df[v_df['ALT_FREQ'] < 0.5].groupby('sample').agg(num_minor_vars=('var', 'nunique'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>GFF_FEATURE</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>ALT_FREQ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4531</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1601</td>\n",
       "      <td>C</td>\n",
       "      <td>-TTA</td>\n",
       "      <td>0.045389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>4531</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6696</td>\n",
       "      <td>C</td>\n",
       "      <td>+T</td>\n",
       "      <td>0.041880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>4531</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>13940</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.057012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>4531</td>\n",
       "      <td>cds-YP_009724390.1</td>\n",
       "      <td>22165</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.030303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>4531</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22583</td>\n",
       "      <td>G</td>\n",
       "      <td>+T</td>\n",
       "      <td>0.092683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>4531</td>\n",
       "      <td>cds-YP_009724390.1</td>\n",
       "      <td>24072</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.039066</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample         GFF_FEATURE    POS REF   ALT  ALT_FREQ\n",
       "5     4531                 NaN   1601   C  -TTA  0.045389\n",
       "8     4531                 NaN   6696   C    +T  0.041880\n",
       "11    4531  cds-YP_009724389.1  13940   A     G  0.057012\n",
       "19    4531  cds-YP_009724390.1  22165   T     C  0.030303\n",
       "22    4531                 NaN  22583   G    +T  0.092683\n",
       "27    4531  cds-YP_009724390.1  24072   A     G  0.039066"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v_df.loc[(v_df['sample']==4531)&(v_df['ALT_FREQ']<0.5)][cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>GFF_FEATURE</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>ALT_FREQ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>4690</td>\n",
       "      <td>cds-YP_009724397.2</td>\n",
       "      <td>29348</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.522132</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample         GFF_FEATURE    POS REF ALT  ALT_FREQ\n",
       "58    4690  cds-YP_009724397.2  29348   G   T  0.522132"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v_df.loc[v_df['POS']==29348][cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>GFF_FEATURE</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>ALT_FREQ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>4531</td>\n",
       "      <td>cds-YP_009724390.1</td>\n",
       "      <td>22343</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>4531</td>\n",
       "      <td>cds-YP_009724391.1</td>\n",
       "      <td>25563</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>4531</td>\n",
       "      <td>cds-YP_009724397.2</td>\n",
       "      <td>28899</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.999853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>4690</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>4806</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.047414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>4690</td>\n",
       "      <td>cds-YP_009724391.1</td>\n",
       "      <td>25563</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>4690</td>\n",
       "      <td>cds-YP_009724397.2</td>\n",
       "      <td>29348</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.522132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>4691</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>7805</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.032258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>4691</td>\n",
       "      <td>cds-YP_009724390.1</td>\n",
       "      <td>22775</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.032258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>4691</td>\n",
       "      <td>cds-YP_009724391.1</td>\n",
       "      <td>25563</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample         GFF_FEATURE    POS REF ALT  ALT_FREQ\n",
       "21    4531  cds-YP_009724390.1  22343   G   T  1.000000\n",
       "28    4531  cds-YP_009724391.1  25563   G   T  1.000000\n",
       "29    4531  cds-YP_009724397.2  28899   G   T  0.999853\n",
       "15    4690  cds-YP_009724389.1   4806   G   T  0.047414\n",
       "50    4690  cds-YP_009724391.1  25563   G   T  1.000000\n",
       "58    4690  cds-YP_009724397.2  29348   G   T  0.522132\n",
       "9     4691  cds-YP_009724389.1   7805   G   T  0.032258\n",
       "22    4691  cds-YP_009724390.1  22775   G   T  0.032258\n",
       "26    4691  cds-YP_009724391.1  25563   G   T  1.000000"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v_df[(v_df['REF']=='G')&(v_df['ALT']=='T')][cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "ans.to_excel('common_variants.xlsx', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(140796, 40)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>REGION</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>REF_DP</th>\n",
       "      <th>REF_RV</th>\n",
       "      <th>REF_QUAL</th>\n",
       "      <th>ALT_DP</th>\n",
       "      <th>ALT_RV</th>\n",
       "      <th>ALT_QUAL</th>\n",
       "      <th>...</th>\n",
       "      <th>REF_RV_54idSEARCH0513SAN_L001_L002_L003_L004.tsv</th>\n",
       "      <th>REF_QUAL_54idSEARCH0513SAN_L001_L002_L003_L004.tsv</th>\n",
       "      <th>ALT_DP_54idSEARCH0513SAN_L001_L002_L003_L004.tsv</th>\n",
       "      <th>ALT_RV_54idSEARCH0513SAN_L001_L002_L003_L004.tsv</th>\n",
       "      <th>ALT_QUAL_54idSEARCH0513SAN_L001_L002_L003_L004.tsv</th>\n",
       "      <th>ALT_FREQ_54idSEARCH0513SAN_L001_L002_L003_L004.tsv</th>\n",
       "      <th>TOTAL_DP_54idSEARCH0513SAN_L001_L002_L003_L004.tsv</th>\n",
       "      <th>PVAL_54idSEARCH0513SAN_L001_L002_L003_L004.tsv</th>\n",
       "      <th>PASS_54idSEARCH0513SAN_L001_L002_L003_L004.tsv</th>\n",
       "      <th>Unnamed: 29</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>241</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>3314.0</td>\n",
       "      <td>769.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>1884</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1541.0</td>\n",
       "      <td>1085.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>1884</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1541.0</td>\n",
       "      <td>1085.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>2455</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>371.0</td>\n",
       "      <td>227.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_045512.2</td>\n",
       "      <td>2455</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>371.0</td>\n",
       "      <td>227.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        REGION   POS REF ALT  REF_DP  REF_RV  REF_QUAL  ALT_DP  ALT_RV  \\\n",
       "0  NC_045512.2   241   C   T     1.0     0.0      72.0  3314.0   769.0   \n",
       "1  NC_045512.2  1884   C   T  1541.0  1085.0      39.0    66.0    35.0   \n",
       "2  NC_045512.2  1884   C   T  1541.0  1085.0      39.0    66.0    35.0   \n",
       "3  NC_045512.2  2455   C   T     0.0     0.0       0.0   371.0   227.0   \n",
       "4  NC_045512.2  2455   C   T     0.0     0.0       0.0   371.0   227.0   \n",
       "\n",
       "   ALT_QUAL  ...  REF_RV_54idSEARCH0513SAN_L001_L002_L003_L004.tsv  \\\n",
       "0      52.0  ...                                               NaN   \n",
       "1      38.0  ...                                               NaN   \n",
       "2      38.0  ...                                               NaN   \n",
       "3      47.0  ...                                               NaN   \n",
       "4      47.0  ...                                               NaN   \n",
       "\n",
       "   REF_QUAL_54idSEARCH0513SAN_L001_L002_L003_L004.tsv  \\\n",
       "0                                                NaN    \n",
       "1                                                NaN    \n",
       "2                                                NaN    \n",
       "3                                                NaN    \n",
       "4                                                NaN    \n",
       "\n",
       "   ALT_DP_54idSEARCH0513SAN_L001_L002_L003_L004.tsv  \\\n",
       "0                                               NaN   \n",
       "1                                               NaN   \n",
       "2                                               NaN   \n",
       "3                                               NaN   \n",
       "4                                               NaN   \n",
       "\n",
       "  ALT_RV_54idSEARCH0513SAN_L001_L002_L003_L004.tsv  \\\n",
       "0                                              NaN   \n",
       "1                                              NaN   \n",
       "2                                              NaN   \n",
       "3                                              NaN   \n",
       "4                                              NaN   \n",
       "\n",
       "  ALT_QUAL_54idSEARCH0513SAN_L001_L002_L003_L004.tsv  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2                                                NaN   \n",
       "3                                                NaN   \n",
       "4                                                NaN   \n",
       "\n",
       "  ALT_FREQ_54idSEARCH0513SAN_L001_L002_L003_L004.tsv  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2                                                NaN   \n",
       "3                                                NaN   \n",
       "4                                                NaN   \n",
       "\n",
       "  TOTAL_DP_54idSEARCH0513SAN_L001_L002_L003_L004.tsv  \\\n",
       "0                                                NaN   \n",
       "1                                                NaN   \n",
       "2                                                NaN   \n",
       "3                                                NaN   \n",
       "4                                                NaN   \n",
       "\n",
       "  PVAL_54idSEARCH0513SAN_L001_L002_L003_L004.tsv  \\\n",
       "0                                            NaN   \n",
       "1                                            NaN   \n",
       "2                                            NaN   \n",
       "3                                            NaN   \n",
       "4                                            NaN   \n",
       "\n",
       "  PASS_54idSEARCH0513SAN_L001_L002_L003_L004.tsv  Unnamed: 29  \n",
       "0                                            NaN          NaN  \n",
       "1                                            NaN          NaN  \n",
       "2                                            NaN          NaN  \n",
       "3                                            NaN          NaN  \n",
       "4                                            NaN          NaN  \n",
       "\n",
       "[5 rows x 40 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "var_fps = glob.glob(f\"/home/gk/analysis/**/variants/illumina/*.tsv\")\n",
    "var_df = (pd.concat((pd.read_csv(f, sep='\\t') for f in var_fps)))\n",
    "print(var_df.shape)\n",
    "var_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'cds-YP_009724389.1': 'ORF1ab',\n",
       " 'cds-YP_009725295.1': 'ORF1ab',\n",
       " 'cds-YP_009724390.1': 'S',\n",
       " 'cds-YP_009724396.1': 'ORF8',\n",
       " 'cds-YP_009724397.2': 'N',\n",
       " 'cds-YP_009724395.1': 'ORF7a',\n",
       " 'cds-YP_009724391.1': 'ORF3a',\n",
       " 'cds-YP_009724393.1': 'M',\n",
       " 'cds-YP_009724394.1': 'ORF6',\n",
       " 'cds-YP_009725255.1': 'ORF10',\n",
       " 'cds-YP_009725318.1': 'ORF7b',\n",
       " 'cds-YP_009724392.1': 'E'}"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_filepath = '/home/al/data/hcov19/gff/NC_045512.2.gff3'\n",
    "fn = gffutils.example_filename(gff_filepath)\n",
    "db = gffutils.create_db(fn, dbfn='test.db', merge_strategy='merge', force=True)\n",
    "all_gffs = var_df['GFF_FEATURE'].dropna().unique()\n",
    "gff2gene = {}\n",
    "for gff in all_gffs:\n",
    "    if gff:\n",
    "        gff2gene[gff] = db[gff]['gene'][0]\n",
    "gff2gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "var_df['gene'] = var_df['GFF_FEATURE'].apply(lambda x: gff2gene.get(x, 'nan'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "# var_df.drop(columns=['GFF_FEATURE'], inplace=True)\n",
    "# try:\n",
    "#     var_df['gene'] = var_df['GFF_FEATURE'].apply(lambda x: gff2gene.get(x, 'nan'))\n",
    "# except:\n",
    "#     raise KeyError('GFF_FEATURE column not found in the input dataframe.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "# var_df.loc[var_df['gene']=='nan']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "# var_df.loc[var_df['gene']=='nan', 'gene'] = var_df.loc[var_df['gene']=='nan', 'POS'].apply(map_gene_to_pos)\n",
    "# var_df.loc[var_df['gene']=='nan']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def map_gene_to_pos(x):\n",
    "    pos = x\n",
    "    if pos >= 269 and pos <= 21555:\n",
    "        return 'ORF1ab'\n",
    "    elif pos >= 21564 and pos <= 25382:\n",
    "        return 'S'\n",
    "    elif pos >= 25410 and pos <= 26214:\n",
    "        return 'ORF3a'\n",
    "    elif pos >= 26247 and pos <= 26471:\n",
    "        return 'E'\n",
    "    elif pos >= 26523 and pos <= 27187:\n",
    "        return 'M'\n",
    "    elif pos >= 27203 and pos <= 27382:\n",
    "        return 'ORF6'\n",
    "    elif pos >= 27398 and pos <= 27754:\n",
    "        return 'ORF7a'\n",
    "    elif pos >= 27757 and pos <= 27887:\n",
    "        return 'ORF7b'\n",
    "    elif pos >= 27896 and pos <= 28257:\n",
    "        return 'ORF8'\n",
    "    elif pos >= 28289 and pos <= 29528:\n",
    "        return 'N'\n",
    "    elif pos >= 29564 and pos <= 29670:\n",
    "        return 'ORF10'\n",
    "    return 'nan'\n",
    "# v_df = v_df.apply(map_gene_to_pos, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.colors as mcolors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "colors = list(mcolors.TABLEAU_COLORS.keys())\n",
    "len(colors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "colors = colors + ['black']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['tab:blue',\n",
       " 'tab:orange',\n",
       " 'tab:green',\n",
       " 'tab:red',\n",
       " 'tab:purple',\n",
       " 'tab:brown',\n",
       " 'tab:pink',\n",
       " 'tab:gray',\n",
       " 'tab:olive',\n",
       " 'tab:cyan',\n",
       " 'black']"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['nan', 'ORF1ab', 'S', 'ORF3a', 'N', 'M', 'ORF7b', 'E'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v_df['gene'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = ['ORF1ab', 'S', 'ORF3a', 'E', 'M', 'ORF6', 'ORF7a',\n",
    "         'ORF7b', 'ORF8', 'N', 'ORF10']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene2color = {}\n",
    "for i, g in enumerate(genes):\n",
    "    gene2color[g] = colors[i]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>GFF_FEATURE</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>ALT_FREQ</th>\n",
       "      <th>var</th>\n",
       "      <th>gene</th>\n",
       "      <th>color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4531</td>\n",
       "      <td>NaN</td>\n",
       "      <td>241</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>241:T</td>\n",
       "      <td>nan</td>\n",
       "      <td>tab:white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4531</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>1059</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1059:T</td>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>tab:blue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4531</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>1102</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1102:T</td>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>tab:blue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4531</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1601</td>\n",
       "      <td>C</td>\n",
       "      <td>-TTA</td>\n",
       "      <td>0.045389</td>\n",
       "      <td>1601:-TTA</td>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>tab:blue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>4531</td>\n",
       "      <td>cds-YP_009724389.1</td>\n",
       "      <td>3037</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3037:T</td>\n",
       "      <td>ORF1ab</td>\n",
       "      <td>tab:blue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>4691</td>\n",
       "      <td>cds-YP_009724390.1</td>\n",
       "      <td>23403</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.999815</td>\n",
       "      <td>23403:G</td>\n",
       "      <td>S</td>\n",
       "      <td>tab:orange</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>4691</td>\n",
       "      <td>cds-YP_009724391.1</td>\n",
       "      <td>25563</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>25563:T</td>\n",
       "      <td>ORF3a</td>\n",
       "      <td>tab:green</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>4691</td>\n",
       "      <td>cds-YP_009724392.1</td>\n",
       "      <td>26300</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>26300:C</td>\n",
       "      <td>E</td>\n",
       "      <td>tab:red</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>4691</td>\n",
       "      <td>NaN</td>\n",
       "      <td>28969</td>\n",
       "      <td>C</td>\n",
       "      <td>-A</td>\n",
       "      <td>0.034876</td>\n",
       "      <td>28969:-A</td>\n",
       "      <td>N</td>\n",
       "      <td>tab:cyan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>4691</td>\n",
       "      <td>NaN</td>\n",
       "      <td>29844</td>\n",
       "      <td>A</td>\n",
       "      <td>+T</td>\n",
       "      <td>0.032520</td>\n",
       "      <td>29844:+T</td>\n",
       "      <td>nan</td>\n",
       "      <td>tab:white</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample         GFF_FEATURE    POS REF   ALT  ALT_FREQ        var    gene  \\\n",
       "0     4531                 NaN    241   C     T  1.000000      241:T     nan   \n",
       "1     4531  cds-YP_009724389.1   1059   C     T  1.000000     1059:T  ORF1ab   \n",
       "3     4531  cds-YP_009724389.1   1102   C     T  1.000000     1102:T  ORF1ab   \n",
       "5     4531                 NaN   1601   C  -TTA  0.045389  1601:-TTA  ORF1ab   \n",
       "6     4531  cds-YP_009724389.1   3037   C     T  1.000000     3037:T  ORF1ab   \n",
       "..     ...                 ...    ...  ..   ...       ...        ...     ...   \n",
       "25    4691  cds-YP_009724390.1  23403   A     G  0.999815    23403:G       S   \n",
       "26    4691  cds-YP_009724391.1  25563   G     T  1.000000    25563:T   ORF3a   \n",
       "27    4691  cds-YP_009724392.1  26300   T     C  0.083333    26300:C       E   \n",
       "28    4691                 NaN  28969   C    -A  0.034876   28969:-A       N   \n",
       "29    4691                 NaN  29844   A    +T  0.032520   29844:+T     nan   \n",
       "\n",
       "         color  \n",
       "0    tab:white  \n",
       "1     tab:blue  \n",
       "3     tab:blue  \n",
       "5     tab:blue  \n",
       "6     tab:blue  \n",
       "..         ...  \n",
       "25  tab:orange  \n",
       "26   tab:green  \n",
       "27     tab:red  \n",
       "28    tab:cyan  \n",
       "29   tab:white  \n",
       "\n",
       "[100 rows x 9 columns]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v_df['color'] = v_df['gene'].apply(lambda x: gene2color.get(x, 'tab:white'))\n",
    "v_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ORF1ab': 'tab:blue',\n",
       " 'S': 'tab:orange',\n",
       " 'ORF3a': 'tab:green',\n",
       " 'E': 'tab:red',\n",
       " 'M': 'tab:purple',\n",
       " 'ORF6': 'tab:brown',\n",
       " 'ORF7a': 'tab:pink',\n",
       " 'ORF7b': 'tab:gray',\n",
       " 'ORF8': 'tab:olive',\n",
       " 'N': 'tab:cyan',\n",
       " 'ORF10': 'black'}"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene2color"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<matplotlib.lines.Line2D at 0x7f0e8bab3b50>,\n",
       " <matplotlib.lines.Line2D at 0x7f0e8bab3610>,\n",
       " <matplotlib.lines.Line2D at 0x7f0e8bab3ca0>,\n",
       " <matplotlib.lines.Line2D at 0x7f0e8bab37f0>,\n",
       " <matplotlib.lines.Line2D at 0x7f0e8bab3520>,\n",
       " <matplotlib.lines.Line2D at 0x7f0e8bab34f0>,\n",
       " <matplotlib.lines.Line2D at 0x7f0e8bab3430>,\n",
       " <matplotlib.lines.Line2D at 0x7f0e8bab3460>,\n",
       " <matplotlib.lines.Line2D at 0x7f0e8bab3eb0>,\n",
       " <matplotlib.lines.Line2D at 0x7f0e8bab3580>,\n",
       " <matplotlib.lines.Line2D at 0x7f0e8baaaac0>]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_color_legends(del2color: dict) -> list:\n",
    "    legend_elements = [Line2D([0], [0], marker='o', color=x, label=y,\n",
    "                       markerfacecolor=x, markersize=15) for y, x in del2color.items()]\n",
    "    return legend_elements\n",
    "legend = get_color_legends(gene2color)\n",
    "legend"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams[\"figure.figsize\"] = (15,15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[3545.0, 0.030302999999999997],\n",
       " [6696.0, 0.035313],\n",
       " [7735.0, 0.0740741],\n",
       " [7805.0, 0.0322581],\n",
       " [7930.0, 0.043478300000000004],\n",
       " [7953.0, 0.0625],\n",
       " [11074.0, 0.034832199999999994],\n",
       " [22000.0, 0.04],\n",
       " [22120.0, 0.04],\n",
       " [22775.0, 0.0322581],\n",
       " [22776.0, 0.0322581],\n",
       " [26300.0, 0.08333330000000001],\n",
       " [28969.0, 0.034876300000000006]]"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s[['POS', 'ALT_FREQ']].values.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Figure size 1080x1080 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import seaborn as sns\n",
    "samples = v_df['sample'].unique()\n",
    "for s_id in samples:\n",
    "    s = v_df.loc[(v_df['sample']==s_id)&(v_df['ALT_FREQ']<0.5)&(v_df['gene']!='nan')]\n",
    "    # I load the seaborn library only to benefit the nice looking feature\n",
    "    plt.vlines(x=s['POS'], ymin=0, ymax=s['ALT_FREQ'], color=s['color'], alpha=0.4)\n",
    "    plt.scatter(s['POS'], s['ALT_FREQ'], color=s['color'], s=50, alpha=1)\n",
    "    for x in s[['POS', 'ALT_FREQ']].values.tolist():\n",
    "        plt.text(x[0], x[1]+0.001, f\"{int(x[0])}\", ha='center')\n",
    "    plt.legend(handles=legend, loc='upper right')\n",
    "    plt.xlabel('POS')\n",
    "    plt.ylabel('ALT_FREQ')\n",
    "    plt.title(f'Minor Variants of Sample {s_id}')\n",
    "    plt.savefig(f'{s_id}_minor_variants.png')\n",
    "    plt.clf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "v_df[v_df['ALT_FREQ']<0.5].sort_values('POS').to_excel('minor_variants.xlsx', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYcAAAD4CAYAAAAHHSreAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8vihELAAAACXBIWXMAAAsTAAALEwEAmpwYAAAQx0lEQVR4nO3dbYxc113H8e+ftdNuH4LbZqH22sEuskwtENisXIeiClGEY4NwkHjhoJI2QnIiYmgROLLbF+1bMFQQEdkybVADIY5wTbCQYYtoK1SJPGzq1I5rlm7TJF7bNFshOxVZ1Q/582Ku0/GeWe9de2Zndv39SKO9c865d849c70/36e9kZlIktTsR7rdAUlS7zEcJEkFw0GSVDAcJEkFw0GSVFjU7Q60ctttt+XKlSu73Q1Jmjeee+6572XmQLuW15PhsHLlSkZGRrrdDUmaNyLi5XYuz8NKkqSC4SBJKhgOkqSC4SBJKhgOkqRCT16tJEm97Mmjp9kzPMqZc5MsW9LPzk1ruGvdYLe71VaGgyTNwpNHT7P70HEmL14G4PS5SXYfOg6woALCw0qSNAt7hkffDIYrJi9eZs/waJd61BmGgyTNwplzk7Mqn68MB0mahWVL+mdVPl8ZDpI0Czs3raF/cd9VZf2L+9i5aU2XetQZnpCWpFm4ctL5wYPHuHD5DQa9WkmSBI2AePyZVwB44r47utybzvCwkiSpYDhIkgqGgySpYDhIkgqGgySpYDhIkgqGgySpYDhIkgqGgySpYDhIkgqGgySpYDhIkgqGgySpYDhIkgqGgySpYDhIkgqGgySpYDhIkgqGgySpUCscIuLOiBiNiLGI2NWiPiLioar+WESsb6r7w4g4EREvRMTjEfHWdq6AJKn9ZgyHiOgDHgY2A2uBuyNi7ZRmm4HV1Ws7sLeadxD4A2AoM38a6AO2ta33kqSOqLPnsAEYy8wXM/MCcADYOqXNVuDRbHgKWBIRS6u6RUB/RCwC3gacaVPfJUkdUiccBoFTTe/Hq7IZ22TmaeDPgFeAs8D5zPxSqw+JiO0RMRIRIxMTE3X7L0nqgDrhEC3Ksk6biHgXjb2KVcAy4O0R8ZFWH5KZ+zNzKDOHBgYGanRLktQpdcJhHFjR9H455aGh6dr8CvCdzJzIzIvAIeAXrr+7kqS5UCccngVWR8SqiLiFxgnlw1PaHAbuqa5a2kjj8NFZGoeTNkbE2yIigA8DJ9vYf0lSByyaqUFmXoqIHcAwjauNHsnMExFxf1W/DzgCbAHGgNeBe6u6pyPiIPB14BJwFNjfiRWRJLXPjOEAkJlHaARAc9m+pukEHphm3k8Dn76BPkqS5ph3SEuSCoaDJKlgOEiSCoaDJKlgOEiSCoaDJKlgOEiSCoaDJKlgOEiSCoaDJKlgOEiSCoaDJKlgOEiSCoaDJKlgOEiSCoaDJKlgOEiSCoaDJKlgOEiSCoaDJKlgOEiSCoaDJKlgOEiSCoaDJKlgOEiSCoaDJKlgOEiSCoaDJKlgOEiSCoaDJKlgOEiSCoaDJKlgOEiSCoaDJKlQKxwi4s6IGI2IsYjY1aI+IuKhqv5YRKxvqlsSEQcj4r8i4mRE3NHOFZAktd+M4RARfcDDwGZgLXB3RKyd0mwzsLp6bQf2NtX9JfCvmflTwM8CJ9vQb0lSB9XZc9gAjGXmi5l5ATgAbJ3SZivwaDY8BSyJiKURcSvwIeDzAJl5ITPPta/7kqROqBMOg8CppvfjVVmdNu8DJoC/iYijEfG5iHh7qw+JiO0RMRIRIxMTE7VXQJLUfnXCIVqUZc02i4D1wN7MXAf8H1CcswDIzP2ZOZSZQwMDAzW6JUnqlDrhMA6saHq/HDhTs804MJ6ZT1flB2mEhSSphy2q0eZZYHVErAJOA9uA357S5jCwIyIOAB8AzmfmWYCIOBURazJzFPgw8M229V6aB548epo9w6OcOTfJsiX97Ny0hrvWTT0yK/WWGcMhMy9FxA5gGOgDHsnMExFxf1W/DzgCbAHGgNeBe5sW8fvAYxFxC/DilDppQXvy6Gl2HzrO5MXLAJw+N8nuQ8cBDAj1tDp7DmTmERoB0Fy2r2k6gQemmfd5YOj6uyjNX3uGR98MhismL15mz/Co4aCe5h3SUgedOTc5q3KpVxgOUgctW9I/q3KpVxgOUgft3LSG/sV9V5X1L+5j56Y1XeqRVE+tcw6Srs+V8woPHjzGhctvMOjVSponDAepw+5aN8jjz7wCwBP3+XcnNT94WEmSVDAcJEkFw0GSVDAcJEkFw0GSVDAcJEkFw0GSVDAcJEkFw0GSVDAcJEkFw0GSVDAcJEkFw0GSVDAcJEkFw0GSVDAcJEkFw0GSVFgwT4J78uhp9gyPcubcJMt8FKMk3ZAFEQ5PHj3N7kPHmbx4GYDT5ybZfeg4gAEhSddhQRxW2jM8+mYwXDF58TJ7hke71CNJmt8WRDicOTc5q3JJ0rUtiHBYtqR/VuWSpGtbEOGwc9Ma+hf3XVXWv7iPnZvWdKlHkjS/LYgT0ldOOj948BgXLr/BoFcrSdINWRDhAI2AePyZVwB44r47utwbSZrfFkw4qOS9H5Kul+GwQHnvh6QbsSBOSKvkvR+SbkStcIiIOyNiNCLGImJXi/qIiIeq+mMRsX5KfV9EHI2If25Xx3Vt3vsh6UbMGA4R0Qc8DGwG1gJ3R8TaKc02A6ur13Zg75T6jwMnb7i3qs17PyTdiDp7DhuAscx8MTMvAAeArVPabAUezYangCURsRQgIpYDvwZ8ro391gy890PSjahzQnoQONX0fhz4QI02g8BZ4C+AB4F3XncvNWve+yHNH714ZWGdcIgWZVmnTUT8OvBqZj4XEb90zQ+J2E7jkBS33357jW5pJt77IfW+Xr2ysM5hpXFgRdP75cCZmm0+CPxGRLxE43DUL0fE37X6kMzcn5lDmTk0MDBQs/uSNL/16pWFdcLhWWB1RKyKiFuAbcDhKW0OA/dUVy1tBM5n5tnM3J2ZyzNzZTXflzPzI+1cAUmaz3r1ysIZDytl5qWI2AEMA33AI5l5IiLur+r3AUeALcAY8Dpwb+e6LEkLx7Il/ZxuEQTdvrKw1h3SmXmERgA0l+1rmk7ggRmW8VXgq7PuoSQtYDs3rbnqnAP0xpWF/vkMSeqiXr2y0HCQpC7rxSsL/dtKkqSC4SBJKhgOkqSC4SBJKhgOkqSC4SBJKhgOkqSC4SBJKhgOkqSC4SBJKvjnMyRpGr34hLa5YjhIUgu9+oS2ueJhJUlqoVef0DZXDAdJaqFXn9A2VwwHSWphuiexdfsJbXPFcJCkFnZuWkP/4r6rynrhCW1zxRPSktRCrz6hba4YDpI0jV58Qttc8bCSJKlgOEiSCoaDJKlgOEiSCoaDJKlgOEiSCoaDJKlgOEiSCoaDJKlgOEiSCoaDJKlgOEiSCoaDJKlgOEiSCrXCISLujIjRiBiLiF0t6iMiHqrqj0XE+qp8RUR8JSJORsSJiPh4u1dAktR+M4ZDRPQBDwObgbXA3RGxdkqzzcDq6rUd2FuVXwL+KDPfD2wEHmgxrySpx9TZc9gAjGXmi5l5ATgAbJ3SZivwaDY8BSyJiKWZeTYzvw6Qmd8HTgI3x2OUJGkeqxMOg8CppvfjlL/gZ2wTESuBdcDTrT4kIrZHxEhEjExMTNToliSpU+qEQ7Qoy9m0iYh3AF8EPpGZr7X6kMzcn5lDmTk0MDBQo1uSpE6pEw7jwIqm98uBM3XbRMRiGsHwWGYeuv6uSpLmSp1weBZYHRGrIuIWYBtweEqbw8A91VVLG4HzmXk2IgL4PHAyMz/b1p5Lkjpm0UwNMvNSROwAhoE+4JHMPBER91f1+4AjwBZgDHgduLea/YPA7wDHI+L5quyTmXmkrWshSWqrGcMBoPplfmRK2b6m6QQeaDHf12h9PkKS1MO8Q1qSVDAcJEkFw0GSVDAcJEkFw0GSVDAcJEkFw0GSVDAcJEkFw0GSVDAcJEkFw0GSVDAcJEkFw0GSVDAcJEkFw0GSVDAcJEkFw0GSVDAcJEkFw0GSVDAcJEkFw0GSVDAcJEkFw0GSVDAcJEkFw0GSVDAcJEkFw0GSVDAcJEkFw0GSVDAcJEkFw0GSVDAcJEkFw0GSVDAcJEmFWuEQEXdGxGhEjEXErhb1EREPVfXHImJ93XklSb1nxnCIiD7gYWAzsBa4OyLWTmm2GVhdvbYDe2cxrySpxyyq0WYDMJaZLwJExAFgK/DNpjZbgUczM4GnImJJRCwFVtaYt23u/Orf896JU7z8tVs7sfh56WNnXwNwTLrM72H+mu67a/d3+rGzr/E/AyvgvjvasrwbVSccBoFTTe/HgQ/UaDNYc14AImI7jb0Obr/99hrdKm3+maX84OT565p3oVq71F9GvcDvYf6a7rtr93e6dumtrHv/0rYu80bUCYdoUZY129SZt1GYuR/YDzA0NNSyzUze+8lPXs9skqQp6oTDOLCi6f1y4EzNNrfUmFeS1GPqXK30LLA6IlZFxC3ANuDwlDaHgXuqq5Y2Aucz82zNeSVJPWbGPYfMvBQRO4BhoA94JDNPRMT9Vf0+4AiwBRgDXgfuvda8HVkTSVLbROMCo94yNDSUIyMj3e6GJM0bEfFcZg61a3neIS1JKhgOkqSC4SBJKhgOkqRCT56QjogJ4OVu92OO3QZ8r9ud6CGOxw85FldzPK52ZTx+IjMH2rXQngyHm1FEjLTzSoP5zvH4Icfiao7H1To1Hh5WkiQVDAdJUsFw6B37u92BHuN4/JBjcTXH42odGQ/POUiSCu45SJIKhoMkqWA4dFBEvBQRxyPi+YgYqcreHRH/FhHfqn6+q6n97ogYi4jRiNjUVP7z1XLGIuKhiGj1EKWeExGPRMSrEfFCU1nb1j8i3hIRT1TlT0fEyjldwVmYZiw+ExGnq+3j+YjY0lS3YMcCICJWRMRXIuJkRJyIiI9X5Tfr9jHdeHRvG8lMXx16AS8Bt00p+1NgVzW9C/iTanot8A3gLcAq4NtAX1X3DHAHjSfr/QuwudvrVnP9PwSsB17oxPoDvwfsq6a3AU90e51nORafAf64RdsFPRZVH5cC66vpdwL/Xa33zbp9TDceXdtG3HOYe1uBL1TTXwDuaio/kJk/yMzv0Hg2xoaIWArcmpn/mY1v9dGmeXpaZv4H8L9Titu5/s3LOgh8uFf3qqYZi+ks6LEAyMyzmfn1avr7wEkaz5y/WbeP6cZjOh0fD8OhsxL4UkQ8FxHbq7Ifz8ZT8qh+/lhVPgicapp3vCobrKanls9X7Vz/N+fJzEvAeeA9Het5Z+yIiGPVYacrh1BuqrGoDm+sA57G7WPqeECXthHDobM+mJnrgc3AAxHxoWu0bZXgeY3yheZ61n++j81e4CeBnwPOAn9eld80YxER7wC+CHwiM1+7VtMWZQtuTFqMR9e2EcOhgzLzTPXzVeAfgQ3Ad6tdP6qfr1bNx4EVTbMvB85U5ctblM9X7Vz/N+eJiEXAj1L/0E3XZeZ3M/NyZr4B/DWN7QNukrGIiMU0fhE+lpmHquKbdvtoNR7d3EYMhw6JiLdHxDuvTAO/CrwAHAY+WjX7KPBP1fRhYFt1RcEqYDXwTLVr/f2I2FgdH7ynaZ75qJ3r37ys3wK+XB1nnReu/BKs/CaN7QNugrGo+v954GRmfrap6qbcPqYbj65uI90+S79QX8D7aFxN8A3gBPCpqvw9wL8D36p+vrtpnk/RuOpglKYrkoChaqP4NvBXVHe29/oLeJzGrvBFGv9r+d12rj/wVuAfaJyMewZ4X7fXeZZj8bfAceBY9Q936c0wFlV/f5HGIY1jwPPVa8tNvH1MNx5d20b88xmSpIKHlSRJBcNBklQwHCRJBcNBklQwHCRJBcNBklQwHCRJhf8Hdsre3qZDLvEAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "my_range=range(1,len(s.index)+1)\n",
    " \n",
    "# Make the plot\n",
    "plt.stem(s['POS'], s['ALT_FREQ'], label=s['gene'])\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
